diff -up wrk/src/threads/ATL_thread_start.c.wrk wrk/src/threads/ATL_thread_start.c
--- wrk/src/threads/ATL_thread_start.c.wrk	2013-09-23 13:46:51.881085276 +0200
+++ wrk/src/threads/ATL_thread_start.c	2013-09-24 16:13:59.021065418 +0200
@@ -101,9 +101,10 @@ int ATL_thread_start(ATL_thread_t *thr,
       ATL_assert(!pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED));
    pthread_attr_setscope(&attr, PTHREAD_SCOPE_SYSTEM); /* no chk, OK to fail */
    #ifdef ATL_PAFF_SETAFFNP
-      CPU_ZERO(&cpuset);
-      CPU_SET(affID, &cpuset);
-      ATL_assert(!pthread_attr_setaffinity_np(&attr, sizeof(cpuset), &cpuset));
+	//affinity crashes a machine with fewer processors than the builder
+      //CPU_ZERO(&cpuset);
+      //CPU_SET(affID, &cpuset);
+      //ATL_assert(!pthread_attr_setaffinity_np(&attr, sizeof(cpuset), &cpuset));
    #elif defined(ATL_PAFF_SETPROCNP)
       ATL_assert(!pthread_attr_setprocessor_np(&attr, (pthread_spu_t)affID,
                                                PTHREAD_BIND_FORCED_NP));
