--- ./ATLAS.first/tune/blas/level3/invtrsm.c	2013-10-22 19:35:03.000000000 +0000
+++ ./ATLAS/tune/blas/level3/invtrsm.c	2013-10-23 21:24:01.000000000 +0000
@@ -525,6 +525,7 @@
    a = A = malloc(i * ATL_MulBySize(incA));
    if (A)
    {
+     memset(A,0,i*ATL_MulBySize(incA)); /* wms  (!!) malloc call above returns non-initialized memory. */
       if (Uplo == TestGE)
          for (i=0; i < k; i++)
             Mjoin(PATL,gegen)(N, N, A+i*incA, lda, N+lda);
