diff -Naur ATLAS.orig/include/atlas_pca.h ATLAS/include/atlas_pca.h
--- ATLAS.orig/include/atlas_pca.h	2013-01-08 19:15:40.000000000 +0100
+++ ATLAS/include/atlas_pca.h	2014-10-23 13:45:36.956698637 +0200
@@ -26,7 +26,7 @@
    #endif
 #elif defined(ATL_ARCH_POWER3) || defined(ATL_ARCH_POWER4) || \
       defined(ATL_ARCH_POWER5) || defined(ATL_ARCH_POWER6) || \
-      defined(ATL_ARCH_POWER7)
+      defined(ATL_ARCH_POWER7) || 1
    #ifdef __GNUC__
       #define ATL_membarrier __asm__ __volatile__ ("dcs")
 /*      #define ATL_USEPCA 1 */
