From fe06dee8f346a8d8ded338bb5080c4cd3b230eef Mon Sep 17 00:00:00 2001
From: Mark Reynolds <mreynolds@redhat.com>
Date: Fri, 5 May 2017 18:33:36 -0400
Subject: [PATCH] Ticket 49231 - Fix backport issue (part2)

---
 ldap/servers/slapd/saslbind.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/ldap/servers/slapd/saslbind.c b/ldap/servers/slapd/saslbind.c
index 8d23c52..75b83fe 100644
--- a/ldap/servers/slapd/saslbind.c
+++ b/ldap/servers/slapd/saslbind.c
@@ -773,7 +773,7 @@ char **ids_sasl_listmech(Slapi_PBlock *pb)
             /* merge into result set */
             dupstr = slapi_ch_strdup(str);
             others = slapi_str2charray_ext(dupstr, ",", 0 /* don't list duplicate mechanisms */);
-            charray_merge(&ret, others, 1);
+            charray_merge(&sup_ret, others, 1);
             charray_free(others);
             slapi_ch_free((void**)&dupstr);
         }
-- 
2.9.3

