From 94ee0c3faaaf6a225f605e00ff623ba9792cedfc Mon Sep 17 00:00:00 2001
From: Miroslav Rezanina <mrezanin@redhat.com>
Date: Wed, 22 May 2013 12:23:00 +0200
Subject: Disable HPET device

Bugzilla: 947441

We do not support HPET, disabling it.

Signed-off-by: Miroslav Rezanina <mrezanin@redhat.com>

diff --git a/default-configs/i386-softmmu.mak b/default-configs/i386-softmmu.mak
index fb6caaf..cf6ee9f 100644
--- a/default-configs/i386-softmmu.mak
+++ b/default-configs/i386-softmmu.mak
@@ -21,7 +21,6 @@ CONFIG_I8257=y
 CONFIG_IDE_ISA=y
 CONFIG_IDE_PIIX=y
 CONFIG_PIIX_PCI=y
-CONFIG_HPET=y
 CONFIG_I8259=y
 CONFIG_PFLASH_CFI01=y
 CONFIG_TPM_TIS=$(CONFIG_TPM)
diff --git a/default-configs/x86_64-softmmu.mak b/default-configs/x86_64-softmmu.mak
index 69cdda0..4234f00 100644
--- a/default-configs/x86_64-softmmu.mak
+++ b/default-configs/x86_64-softmmu.mak
@@ -21,7 +21,6 @@ CONFIG_I8257=y
 CONFIG_IDE_ISA=y
 CONFIG_IDE_PIIX=y
 CONFIG_PIIX_PCI=y
-CONFIG_HPET=y
 CONFIG_I8259=y
 CONFIG_PFLASH_CFI01=y
 CONFIG_TPM_TIS=$(CONFIG_TPM)
