From d74dbb4a6f6531bfb3b354ac2545882d7558a55a Mon Sep 17 00:00:00 2001
From: Igor Gnatenko <i.gnatenko.brain@gmail.com>
Date: Thu, 25 Aug 2016 21:49:55 +0200
Subject: [PATCH] util: detect armv7hnl properly

Signed-off-by: Igor Gnatenko <i.gnatenko.brain@gmail.com>
---
 src/util.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/src/util.c b/src/util.c
index 0e9352c..8e41226 100644
--- a/src/util.c
+++ b/src/util.c
@@ -136,10 +136,11 @@ hy_detect_arch(char **arch)
 	int ret = parse_cpu_flags(&flags, "Features");
 	if (ret)
 	    return ret;
-	if (flags & (ARM_NEON | ARM_VFP3))
-	    strcpy(un.machine, "armv7hnl");
-	else if (flags & ARM_VFP3)
-	    strcpy(un.machine, "armv7hl");
+        if (flags & ARM_VFP3)
+            if (flags & ARM_NEON)
+                strcpy(un.machine, "armv7hnl");
+            else
+                strcpy(un.machine, "armv7hl");
     }
 #ifdef __MIPSEL__
     if (!strcmp(un.machine, "mips"))
-- 
2.9.3

