From a0811c1ad96236fefc8eebe7a79a9b86ee654892 Mon Sep 17 00:00:00 2001
From: Petr Holasek <pholasek@redhat.com>
Date: Thu, 19 Mar 2015 14:58:35 +0100
Subject: [PATCH 1/1] irqbalance: set IRQBALANCE_DEBUG variable implies
 foreground mode

Signed-off-by: Petr Holasek <pholasek@redhat.com>
---
 irqbalance.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/irqbalance.c b/irqbalance.c
index 29967ba..226fcc6 100644
--- a/irqbalance.c
+++ b/irqbalance.c
@@ -269,8 +269,10 @@ int main(int argc, char** argv)
 	if (getenv("IRQBALANCE_ONESHOT")) 
 		one_shot_mode=1;
 
-	if (getenv("IRQBALANCE_DEBUG")) 
+	if (getenv("IRQBALANCE_DEBUG")) {
 		debug_mode=1;
+		foreground_mode=1;
+	}
 
 	/*
  	 * If we are't in debug mode, don't dump anything to the console
-- 
2.12.2

