From 854425506dc26d1e118122b2f87f1b7f5641ee35 Mon Sep 17 00:00:00 2001
From: Petr Holasek <pholasek@redhat.com>
Date: Wed, 10 Dec 2014 15:12:10 +0100
Subject: [PATCH] removing unused variable cache_stat

Signed-off-by: Petr Holasek <pholasek@redhat.com>
---
 cputree.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/cputree.c b/cputree.c
index caedbb6..c56b1be 100644
--- a/cputree.c
+++ b/cputree.c
@@ -234,7 +234,6 @@ static void do_one_cpu(char *path)
 	cpu_set(cpu->number, cache_mask);
 	max_cache_index = 0;
 	cache_index = 1;
-	cache_stat = 0;
 	do {
 		struct stat sb;
 		snprintf(new_path, PATH_MAX, "%s/cache/index%d/shared_cpu_map", path, cache_index);
-- 
2.1.0

