diff -up irqbalance-1.0.7/irqbalance.1.orig irqbalance-1.0.7/irqbalance.1
--- irqbalance-1.0.7/irqbalance.1.orig	2014-08-19 14:42:45.760710296 +0200
+++ irqbalance-1.0.7/irqbalance.1	2014-08-19 14:43:00.246715877 +0200
@@ -57,6 +57,8 @@ irq is balanced, but the assigned object
 .P
 .I ignore
 irq affinity hint value is completely ignored
+.P
+The default value for hintpolicy is ignore
 
 .TP
 .B -p, --powerthresh=<threshold>
diff -up irqbalance-1.0.7/irqbalance.c.orig irqbalance-1.0.7/irqbalance.c
--- irqbalance-1.0.7/irqbalance.c.orig	2013-10-10 17:14:07.000000000 +0200
+++ irqbalance-1.0.7/irqbalance.c	2014-08-19 14:43:00.247715877 +0200
@@ -48,7 +48,7 @@ int foreground_mode;
 int numa_avail;
 int need_rescan;
 unsigned int log_mask = TO_ALL;
-enum hp_e hint_policy = HINT_POLICY_SUBSET;
+enum hp_e hint_policy = HINT_POLICY_IGNORE;
 unsigned long power_thresh = ULONG_MAX;
 unsigned long deepest_cache = ULONG_MAX;
 unsigned long long cycle_count = 0;
