diff --git a/tests/backtrace-subr.sh b/tests/backtrace-subr.sh
index c1f3156..477df0b 100644
--- a/tests/backtrace-subr.sh
+++ b/tests/backtrace-subr.sh
@@ -105,6 +105,12 @@ check_native_unsupported()
 	exit 77
       fi
     ;;
+    x86_64 )
+      if egrep 'backtrace: backtrace.c:114: callback_verify: Assertion' $err; then
+	echo >&2 $testname: temp override for x86_64
+	exit 77
+      fi
+    ;;
   esac
 }
 
