diff --git a/libasm/Makefile.am b/libasm/Makefile.am
index b15ee55..0087077 100644
--- a/libasm/Makefile.am
+++ b/libasm/Makefile.am
@@ -31,7 +31,7 @@ include $(top_srcdir)/config/eu.am
 AM_CPPFLAGS += -I$(top_srcdir)/libelf -I$(top_srcdir)/libebl -I$(top_srcdir)/libdw -I$(top_srcdir)/libdwelf
 
 GCC_INCLUDE = -I$(shell $(CC) -print-file-name=include)
-VERSION = 1
+VERSION = dts.1
 
 lib_LIBRARIES = libasm.a
 noinst_LIBRARIES = libasm_pic.a
diff --git a/libdw/Makefile.am b/libdw/Makefile.am
index 4cb3671..bc46a13 100644
--- a/libdw/Makefile.am
+++ b/libdw/Makefile.am
@@ -32,7 +32,7 @@ if BUILD_STATIC
 AM_CFLAGS += -fPIC
 endif
 AM_CPPFLAGS += -I$(srcdir)/../libelf
-VERSION = 1
+VERSION = dts.1
 
 lib_LIBRARIES = libdw.a
 noinst_LIBRARIES = libdw_pic.a libdw_static_pic.a
diff --git a/libelf/Makefile.am b/libelf/Makefile.am
index 0c74e3f..fdf98a4 100644
--- a/libelf/Makefile.am
+++ b/libelf/Makefile.am
@@ -32,7 +32,7 @@ if BUILD_STATIC
 AM_CFLAGS += -fPIC
 endif
 GCC_INCLUDE = -I$(shell $(CC) -print-file-name=include)
-VERSION = 1
+VERSION = dts.1
 
 lib_LIBRARIES = libelf.a
 noinst_LIBRARIES = libelf_pic.a
