--- pyparted-3.9/Makefile.orig	2014-02-13 10:00:59.129864074 -0500
+++ pyparted-3.9/Makefile	2014-02-13 10:13:52.547882402 -0500
@@ -37,8 +37,11 @@
 	@$(PYTHON) setup.py build
 
 test: all
+	@which coverage || (echo "*** Please install python-coverage ***"; exit 2)
+	@echo "*** Running unittests with coverage ***"
 	@env PYTHONPATH=$$(find $$(pwd) -name "*.so" | head -n 1 | xargs dirname) \
-	$(PYTHON) -m unittest discover -v
+	coverage run --branch -m unittest discover -v
+	coverage report --include='build/*'
 
 check: all
 	env PYTHONPATH=$$(find $$(pwd) -name "*.so" | head -n 1 | xargs dirname) \
