diff -ruN pyparted-3.9-orig/src/parted/__init__.py pyparted-3.9/src/parted/__init__.py
--- pyparted-3.9-orig/src/parted/__init__.py	2015-03-27 20:44:41.903949352 +0100
+++ pyparted-3.9/src/parted/__init__.py	2015-03-27 20:49:10.168220174 +0100
@@ -278,6 +278,9 @@
               'ppc64': ['msdos', 'mac', 'amiga', 'gpt'],
               'ppc64le': ['msdos', 'gpt'],
               'x86_64': ['msdos', 'gpt'],
+              'armv5tel': ['msdos', 'gpt'],
+              'armv6l': ['msdos', 'gpt'],
+              'armv7l': ['msdos', 'gpt'],
               'aarch64': ['msdos', 'gpt']}
 
 # Adapted from:
@@ -312,9 +315,9 @@
 __archLabels = (('amiga', 'ppc(64)?$'),
                 ('bsd', 'alpha$'),
                 ('dasd', 's390x?$'),
-                ('gpt', 'i[3-6]86$|x86_64$|ia64$|ppc(64|64le)?$|aarch64$'),
+                ('gpt', 'i[3-6]86$|x86_64$|ia64$|ppc(64|64le)?$|aarch64$|armv[5-7]'),
                 ('mac', 'ppc(64)?$'),
-                ('msdos', 'i[3-6]86$|x86_64$|s390x?$|alpha$|ia64$|ppc(64|64le)?$|aarch64$'),
+                ('msdos', 'i[3-6]86$|x86_64$|s390x?$|alpha$|ia64$|ppc(64|64le)?$|aarch64$|armv[5-7]'),
                 ('sun', 'sparc(64)?$'))
 
 def getLabels(arch=None):
diff -ruN pyparted-3.9-orig/tests/test_parted_parted.py pyparted-3.9/tests/test_parted_parted.py
--- pyparted-3.9-orig/tests/test_parted_parted.py	2015-03-27 20:44:41.903949352 +0100
+++ pyparted-3.9/tests/test_parted_parted.py	2015-03-27 20:50:07.053064146 +0100
@@ -60,6 +60,9 @@
         self.assertSetEqual(parted.getLabels('ppc64le'), {'gpt', 'msdos'})
         self.assertSetEqual(parted.getLabels('alpha'), {'bsd', 'msdos'})
         self.assertSetEqual(parted.getLabels('ia64'), {'gpt', 'msdos'})
+        self.assertSetEqual(parted.getLabels('armv5tel'), {'gpt', 'msdos'})
+        self.assertSetEqual(parted.getLabels('armv6l'), {'gpt', 'msdos'})
+        self.assertSetEqual(parted.getLabels('armv7l'), {'gpt', 'msdos'})
         self.assertSetEqual(parted.getLabels('aarch64'), {'gpt', 'msdos'})
 
 class GetDeviceTestCase(RequiresDeviceNode):
