diff -up pyparted-3.9/src/parted/__init__.py.aarch64 pyparted-3.9/src/parted/__init__.py
--- pyparted-3.9/src/parted/__init__.py.aarch64	2014-01-23 17:28:28.370005898 -0500
+++ pyparted-3.9/src/parted/__init__.py	2014-01-23 17:57:51.257702523 -0500
@@ -275,7 +275,8 @@ archLabels = {'i386': ['msdos', 'gpt'],
               'sparc': ['sun'],
               'ia64': ['msdos', 'gpt'],
               'ppc': ['msdos', 'mac', 'amiga', 'gpt'],
-              'x86_64': ['msdos', 'gpt']}
+              'x86_64': ['msdos', 'gpt'],
+              'aarch64': ['msdos', 'gpt']}
 
 # Adapted from:
 # http://stackoverflow.com/questions/922550/how-to-mark-a-global-as-deprecated-in-python
@@ -309,9 +310,9 @@ def Deprecated(mod, deprecated={}):
 __archLabels = (('amiga', 'ppc(64)?$'),
                 ('bsd', 'alpha$'),
                 ('dasd', 's390x?$'),
-                ('gpt', 'i[3-6]86$|x86_64$|ia64$|ppc(64)?$'),
+                ('gpt', 'i[3-6]86$|x86_64$|ia64$|ppc(64)?$|aarch64$'),
                 ('mac', 'ppc(64)?$'),
-                ('msdos', 'i[3-6]86$|x86_64$|s390x?$|alpha$|ia64$|ppc(64)?$'),
+                ('msdos', 'i[3-6]86$|x86_64$|s390x?$|alpha$|ia64$|ppc(64)?$|aarch64$'),
                 ('sun', 'sparc(64)?$'))
 
 def getLabels(arch=None):
diff -up pyparted-3.9/tests/test_parted_parted.py.aarch64 pyparted-3.9/tests/test_parted_parted.py
--- pyparted-3.9/tests/test_parted_parted.py.aarch64	2011-06-27 11:24:21.000000000 -0400
+++ pyparted-3.9/tests/test_parted_parted.py	2014-01-23 17:59:09.640810266 -0500
@@ -59,6 +59,7 @@ class GetLabelsTestCase(unittest.TestCas
         self.assertSetEqual(parted.getLabels('ppc64'), {'amiga', 'gpt', 'mac', 'msdos'})
         self.assertSetEqual(parted.getLabels('alpha'), {'bsd', 'msdos'})
         self.assertSetEqual(parted.getLabels('ia64'), {'gpt', 'msdos'})
+        self.assertSetEqual(parted.getLabels('aarch64'), {'gpt', 'msdos'})
 
 class GetDeviceTestCase(RequiresDeviceNode):
     def runTest(self):
