From 96f91f23a6a413535cc2f0e8492e2300373fed40 Mon Sep 17 00:00:00 2001
From: Jan Friesse <jfriesse@redhat.com>
Date: Mon, 7 Nov 2016 17:39:12 +0100
Subject: [PATCH] libvotequorum: Bump version

Signed-off-by: Jan Friesse <jfriesse@redhat.com>
---
 lib/libvotequorum.verso |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/lib/libvotequorum.verso b/lib/libvotequorum.verso
index 66ce77b..ae9a76b 100644
--- a/lib/libvotequorum.verso
+++ b/lib/libvotequorum.verso
@@ -1 +1 @@
-7.0.0
+8.0.0
-- 
1.7.1

