Index: tests/test_ssl.py
===================================================================
--- tests/test_ssl.py	(revision 739)
+++ tests/test_ssl.py	(working copy)
@@ -376,7 +376,7 @@
     def test_sslv23_no_v2(self):
         if fips_mode: # TLS is required in FIPS mode
             return
-        self.args.append('-no_tls1')
+        self.args.append('-ssl3')
         pid = self.start_server(self.args)
         try:
             ctx = SSL.Context('sslv23')
@@ -390,7 +390,7 @@
     def test_sslv23_no_v2_no_service(self):
         if fips_mode: # TLS is required in FIPS mode
             return
-        self.args = self.args + ['-no_tls1', '-no_ssl3']
+        self.args = self.args + ['-ssl2']
         pid = self.start_server(self.args)
         try:
             ctx = SSL.Context('sslv23')
@@ -403,7 +403,7 @@
     def test_sslv23_weak_crypto(self):
         if fips_mode: # TLS is required in FIPS mode
             return
-        self.args = self.args + ['-no_tls1', '-no_ssl3']
+        self.args = self.args + ['-ssl2']
         pid = self.start_server(self.args)
         try:
             ctx = SSL.Context('sslv23', weak_crypto=1)
