--- libquadmath/Makefile.am.jj	2016-04-15 16:53:53.207442732 +0200
+++ libquadmath/Makefile.am	2016-07-21 19:04:01.472619416 +0200
@@ -35,15 +35,19 @@ version_dep =
 endif
 
 toolexeclib_LTLIBRARIES = libquadmath.la
+noinst_LTLIBRARIES = libquadmathconvenience.la
 libquadmath_la_LIBADD = 
 libquadmath_la_LDFLAGS = -version-info `grep -v '^\#' $(srcdir)/libtool-version` \
 			 $(version_arg) $(lt_host_flags) -lm
 libquadmath_la_DEPENDENCIES = $(version_dep) $(libquadmath_la_LIBADD)
+libquadmathconvenience_la_LIBADD =
+libquadmathconvenience_la_LDFLAGS =
+libquadmathconvenience_la_DEPENDENCIES =
 
 nodist_libsubinclude_HEADERS = quadmath.h quadmath_weak.h
 libsubincludedir = $(libdir)/gcc/$(target_alias)/$(gcc_version)/include
 
-libquadmath_la_SOURCES = \
+sources = \
   math/x2y2m1q.c math/isinf_nsq.c math/acoshq.c math/fmodq.c \
   math/acosq.c math/frexpq.c \
   math/rem_pio2q.c math/asinhq.c math/hypotq.c math/remainderq.c \
@@ -69,7 +73,8 @@ libquadmath_la_SOURCES = \
   printf/mul_n.c printf/mul.c printf/printf_fphex.c printf/printf_fp.c \
   printf/quadmath-printf.c printf/rshift.c printf/submul_1.c printf/sub_n.c \
   strtod/strtoflt128.c strtod/mpn2flt128.c strtod/tens_in_limb.c
-
+libquadmath_la_SOURCES = $(sources)
+libquadmathconvenience_la_SOURCES = $(sources)
 
 # Work around what appears to be a GNU make bug handling MAKEFLAGS
 # values defined in terms of make variables, as is the case for CC and
--- libquadmath/Makefile.in.jj	2016-04-15 16:53:53.170443230 +0200
+++ libquadmath/Makefile.in	2016-07-21 19:04:23.626349760 +0200
@@ -54,6 +54,7 @@ build_triplet = @build@
 host_triplet = @host@
 target_triplet = @target@
 @BUILD_LIBQUADMATH_FALSE@libquadmath_la_DEPENDENCIES =
+@BUILD_LIBQUADMATH_FALSE@libquadmathconvenience_la_DEPENDENCIES =
 subdir = .
 DIST_COMMON = ChangeLog $(srcdir)/Makefile.in $(srcdir)/Makefile.am \
 	$(top_srcdir)/configure $(am__configure_deps) \
@@ -108,9 +109,9 @@ am__uninstall_files_from_dir = { \
   }
 am__installdirs = "$(DESTDIR)$(toolexeclibdir)" "$(DESTDIR)$(infodir)" \
 	"$(DESTDIR)$(libsubincludedir)"
-LTLIBRARIES = $(toolexeclib_LTLIBRARIES)
+LTLIBRARIES = $(noinst_LTLIBRARIES) $(toolexeclib_LTLIBRARIES)
 am__dirstamp = $(am__leading_dot)dirstamp
-@BUILD_LIBQUADMATH_TRUE@am_libquadmath_la_OBJECTS = math/x2y2m1q.lo \
+@BUILD_LIBQUADMATH_TRUE@am__objects_1 = math/x2y2m1q.lo \
 @BUILD_LIBQUADMATH_TRUE@	math/isinf_nsq.lo math/acoshq.lo \
 @BUILD_LIBQUADMATH_TRUE@	math/fmodq.lo math/acosq.lo \
 @BUILD_LIBQUADMATH_TRUE@	math/frexpq.lo math/rem_pio2q.lo \
@@ -167,12 +168,22 @@ am__dirstamp = $(am__leading_dot)dirstam
 @BUILD_LIBQUADMATH_TRUE@	printf/sub_n.lo strtod/strtoflt128.lo \
 @BUILD_LIBQUADMATH_TRUE@	strtod/mpn2flt128.lo \
 @BUILD_LIBQUADMATH_TRUE@	strtod/tens_in_limb.lo
+@BUILD_LIBQUADMATH_TRUE@am_libquadmath_la_OBJECTS = $(am__objects_1)
 libquadmath_la_OBJECTS = $(am_libquadmath_la_OBJECTS)
 libquadmath_la_LINK = $(LIBTOOL) --tag=CC $(AM_LIBTOOLFLAGS) \
 	$(LIBTOOLFLAGS) --mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) \
 	$(libquadmath_la_LDFLAGS) $(LDFLAGS) -o $@
 @BUILD_LIBQUADMATH_TRUE@am_libquadmath_la_rpath = -rpath \
 @BUILD_LIBQUADMATH_TRUE@	$(toolexeclibdir)
+@BUILD_LIBQUADMATH_TRUE@am_libquadmathconvenience_la_OBJECTS =  \
+@BUILD_LIBQUADMATH_TRUE@	$(am__objects_1)
+libquadmathconvenience_la_OBJECTS =  \
+	$(am_libquadmathconvenience_la_OBJECTS)
+libquadmathconvenience_la_LINK = $(LIBTOOL) --tag=CC \
+	$(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) --mode=link $(CCLD) \
+	$(AM_CFLAGS) $(CFLAGS) $(libquadmathconvenience_la_LDFLAGS) \
+	$(LDFLAGS) -o $@
+@BUILD_LIBQUADMATH_TRUE@am_libquadmathconvenience_la_rpath =
 DEFAULT_INCLUDES = -I.@am__isrc@
 depcomp = $(SHELL) $(top_srcdir)/../depcomp
 am__depfiles_maybe = depfiles
@@ -186,7 +197,8 @@ CCLD = $(CC)
 LINK = $(LIBTOOL) --tag=CC $(AM_LIBTOOLFLAGS) $(LIBTOOLFLAGS) \
 	--mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) \
 	$(LDFLAGS) -o $@
-SOURCES = $(libquadmath_la_SOURCES)
+SOURCES = $(libquadmath_la_SOURCES) \
+	$(libquadmathconvenience_la_SOURCES)
 MULTISRCTOP = 
 MULTIBUILDTOP = 
 MULTIDIRS = 
@@ -348,14 +360,18 @@ AUTOMAKE_OPTIONS = 1.8 foreign
 @BUILD_LIBQUADMATH_TRUE@@LIBQUAD_USE_SYMVER_GNU_TRUE@@LIBQUAD_USE_SYMVER_TRUE@version_dep = $(srcdir)/quadmath.map
 @BUILD_LIBQUADMATH_TRUE@@LIBQUAD_USE_SYMVER_SUN_TRUE@@LIBQUAD_USE_SYMVER_TRUE@version_dep = quadmath.map-sun
 @BUILD_LIBQUADMATH_TRUE@toolexeclib_LTLIBRARIES = libquadmath.la
+@BUILD_LIBQUADMATH_TRUE@noinst_LTLIBRARIES = libquadmathconvenience.la
 @BUILD_LIBQUADMATH_TRUE@libquadmath_la_LIBADD = 
 @BUILD_LIBQUADMATH_TRUE@libquadmath_la_LDFLAGS = -version-info `grep -v '^\#' $(srcdir)/libtool-version` \
 @BUILD_LIBQUADMATH_TRUE@			 $(version_arg) $(lt_host_flags) -lm
 
 @BUILD_LIBQUADMATH_TRUE@libquadmath_la_DEPENDENCIES = $(version_dep) $(libquadmath_la_LIBADD)
+@BUILD_LIBQUADMATH_TRUE@libquadmathconvenience_la_LIBADD = 
+@BUILD_LIBQUADMATH_TRUE@libquadmathconvenience_la_LDFLAGS = 
+@BUILD_LIBQUADMATH_TRUE@libquadmathconvenience_la_DEPENDENCIES = 
 @BUILD_LIBQUADMATH_TRUE@nodist_libsubinclude_HEADERS = quadmath.h quadmath_weak.h
 @BUILD_LIBQUADMATH_TRUE@libsubincludedir = $(libdir)/gcc/$(target_alias)/$(gcc_version)/include
-@BUILD_LIBQUADMATH_TRUE@libquadmath_la_SOURCES = \
+@BUILD_LIBQUADMATH_TRUE@sources = \
 @BUILD_LIBQUADMATH_TRUE@  math/x2y2m1q.c math/isinf_nsq.c math/acoshq.c math/fmodq.c \
 @BUILD_LIBQUADMATH_TRUE@  math/acosq.c math/frexpq.c \
 @BUILD_LIBQUADMATH_TRUE@  math/rem_pio2q.c math/asinhq.c math/hypotq.c math/remainderq.c \
@@ -382,6 +398,8 @@ AUTOMAKE_OPTIONS = 1.8 foreign
 @BUILD_LIBQUADMATH_TRUE@  printf/quadmath-printf.c printf/rshift.c printf/submul_1.c printf/sub_n.c \
 @BUILD_LIBQUADMATH_TRUE@  strtod/strtoflt128.c strtod/mpn2flt128.c strtod/tens_in_limb.c
 
+@BUILD_LIBQUADMATH_TRUE@libquadmath_la_SOURCES = $(sources)
+@BUILD_LIBQUADMATH_TRUE@libquadmathconvenience_la_SOURCES = $(sources)
 
 # Work around what appears to be a GNU make bug handling MAKEFLAGS
 # values defined in terms of make variables, as is the case for CC and
@@ -504,6 +522,15 @@ $(srcdir)/config.h.in: @MAINTAINER_MODE_
 
 distclean-hdr:
 	-rm -f config.h stamp-h1
+
+clean-noinstLTLIBRARIES:
+	-test -z "$(noinst_LTLIBRARIES)" || rm -f $(noinst_LTLIBRARIES)
+	@list='$(noinst_LTLIBRARIES)'; for p in $$list; do \
+	  dir="`echo $$p | sed -e 's|/[^/]*$$||'`"; \
+	  test "$$dir" != "$$p" || dir=.; \
+	  echo "rm -f \"$${dir}/so_locations\""; \
+	  rm -f "$${dir}/so_locations"; \
+	done
 install-toolexeclibLTLIBRARIES: $(toolexeclib_LTLIBRARIES)
 	@$(NORMAL_INSTALL)
 	@list='$(toolexeclib_LTLIBRARIES)'; test -n "$(toolexeclibdir)" || list=; \
@@ -688,6 +715,8 @@ strtod/tens_in_limb.lo: strtod/$(am__dir
 	strtod/$(DEPDIR)/$(am__dirstamp)
 libquadmath.la: $(libquadmath_la_OBJECTS) $(libquadmath_la_DEPENDENCIES) $(EXTRA_libquadmath_la_DEPENDENCIES) 
 	$(libquadmath_la_LINK) $(am_libquadmath_la_rpath) $(libquadmath_la_OBJECTS) $(libquadmath_la_LIBADD) $(LIBS)
+libquadmathconvenience.la: $(libquadmathconvenience_la_OBJECTS) $(libquadmathconvenience_la_DEPENDENCIES) $(EXTRA_libquadmathconvenience_la_DEPENDENCIES) 
+	$(libquadmathconvenience_la_LINK) $(am_libquadmathconvenience_la_rpath) $(libquadmathconvenience_la_OBJECTS) $(libquadmathconvenience_la_LIBADD) $(LIBS)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT)
@@ -1316,7 +1345,8 @@ maintainer-clean-generic:
 clean: clean-am clean-multi
 
 clean-am: clean-aminfo clean-generic clean-libtool \
-	clean-toolexeclibLTLIBRARIES mostlyclean-am
+	clean-noinstLTLIBRARIES clean-toolexeclibLTLIBRARIES \
+	mostlyclean-am
 
 distclean: distclean-am distclean-multi
 	-rm -f $(am__CONFIG_DISTCLEAN_FILES)
@@ -1487,8 +1517,9 @@ uninstall-am: uninstall-dvi-am uninstall
 
 .PHONY: CTAGS GTAGS all all-am all-local all-multi am--refresh check \
 	check-am clean clean-aminfo clean-generic clean-libtool \
-	clean-multi clean-toolexeclibLTLIBRARIES ctags dist-info \
-	distclean distclean-compile distclean-generic distclean-hdr \
+	clean-multi clean-noinstLTLIBRARIES \
+	clean-toolexeclibLTLIBRARIES ctags dist-info distclean \
+	distclean-compile distclean-generic distclean-hdr \
 	distclean-libtool distclean-multi distclean-tags dvi dvi-am \
 	html html-am info info-am install install-am install-data \
 	install-data-am install-dvi install-dvi-am install-exec \
