diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/locale-inst-asm.S gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/locale-inst-asm.S
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/locale-inst-asm.S	2017-07-03 12:35:38.145446630 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/locale-inst-asm.S	2017-07-03 14:18:18.262272416 +0200
@@ -22,6 +22,11 @@
 
 #ifdef __i386__
 #define ALIGN1	.align 4
+#elif defined __arm__
+#define ALIGN1	.align 4
+#define OBJECT %object
+#define SECTION1(x).section .gnu.linkonce.d.rel.ro.x,"aw",%progbits
+#define SECTION3(x).section .gnu.linkonce.r.x,"a",%progbits
 #elif defined __x86_64__
 #define ALIGN1	.align 32
 #define ALIGN2	.align 16
@@ -288,4 +293,4 @@
 	POINTER	SYM(_ZTISt16__numpunct_cacheIwE)
 	VPTR(_ZNSt16__numpunct_cacheIwED1Ev)
 	VPTR(_ZNSt16__numpunct_cacheIwED0Ev)
-	.section	.note.GNU-stack,"",@progbits
+	.section	.note.GNU-stack,"",%progbits
diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared98/decimal-rtti.S gcc-6.3.1-20170216/libstdc++-v3/src/nonshared98/decimal-rtti.S
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared98/decimal-rtti.S	2017-07-03 14:43:40.511082860 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared98/decimal-rtti.S	2017-07-03 14:14:14.138216542 +0200
@@ -21,9 +21,14 @@
    <http://www.gnu.org/licenses/>.  */
 
 #if defined __x86_64__ || defined __powerpc64__ || defined __s390x__ || defined __ia64__ || defined __aarch64__ \
-    || defined __i386__ || defined __powerpc__ || defined __s390__
+    || defined __i386__ || defined __powerpc__ || defined __s390__ || defined __arm__
 #ifdef __i386__
 #define ALIGN1	.align 4
+#elif defined __arm__
+#define ALIGN1	.align 4
+#define OBJECT %object
+#define SECTION1(x).section .gnu.linkonce.d.rel.ro.x,"aw",%progbits
+#define SECTION3(x).section .gnu.linkonce.r.x,"a",%progbits
 #elif defined __x86_64__
 #define ALIGN1	.align 32
 #define ALIGN2	.align 16
@@ -271,4 +276,4 @@
 _ZTSDf:
 	STRING	"Df"
 #endif
-	.section	.note.GNU-stack,"",@progbits
+	.section	.note.GNU-stack,"",%progbits
diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared98/float128.S gcc-6.3.1-20170216/libstdc++-v3/src/nonshared98/float128.S
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared98/float128.S	2017-07-03 14:43:46.011059614 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared98/float128.S	2017-07-03 14:16:20.767726815 +0200
@@ -20,9 +20,14 @@
    see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
    <http://www.gnu.org/licenses/>.  */
 
-#if defined __x86_64__ || defined __i386__ || defined __ia64__
+#if defined __x86_64__ || defined __i386__ || defined __ia64__ || defined __arm__
 #ifdef __i386__
 #define ALIGN1	.align 4
+#elif defined __arm__
+#define ALIGN1	.align 4
+#define OBJECT %object
+#define SECTION1(x).section .gnu.linkonce.d.rel.ro.x,"aw",%progbits
+#define SECTION3(x).section .gnu.linkonce.r.x,"a",%progbits
 #elif defined __x86_64__
 #define ALIGN1	.align 32
 #define ALIGN2	.align 16
@@ -159,4 +164,4 @@
 _ZTSg:
 	STRING	"g"
 #endif
-	.section	.note.GNU-stack,"",@progbits
+	.section	.note.GNU-stack,"",%progbits
diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared98/int128.S gcc-6.3.1-20170216/libstdc++-v3/src/nonshared98/int128.S
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared98/int128.S	2017-07-03 14:43:51.191037720 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared98/int128.S	2017-07-03 13:49:53.148806566 +0200
@@ -213,4 +213,4 @@
 _ZTSo:
 	STRING	"o"
 #endif
-	.section	.note.GNU-stack,"",@progbits
+	.section	.note.GNU-stack,"",%progbits
diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared98/nullptr-rtti.S gcc-6.3.1-20170216/libstdc++-v3/src/nonshared98/nullptr-rtti.S
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared98/nullptr-rtti.S	2017-07-03 14:43:56.011017348 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared98/nullptr-rtti.S	2017-07-03 13:51:34.473432976 +0200
@@ -21,9 +21,14 @@
    <http://www.gnu.org/licenses/>.  */
 
 #if defined __x86_64__ || defined __powerpc64__ || defined __s390x__ || defined __ia64__ || defined __aarch64__ \
-    || defined __i386__ || defined __powerpc__ || defined __s390__
+    || defined __i386__ || defined __powerpc__ || defined __s390__ || defined __arm__
 #ifdef __i386__
 #define ALIGN1	.align 4
+#elif defined __arm__
+#define ALIGN1	.align 4
+#define OBJECT	%object
+#define SECTION1(x).section .gnu.linkonce.d.rel.ro.x,"aw",%progbits
+#define SECTION3(x).section .gnu.linkonce.r.x,"a",%progbits
 #elif defined __x86_64__
 #define ALIGN1	.align 32
 #define ALIGN2	.align 16
@@ -163,4 +168,4 @@
 _ZTSDn:
 	STRING	"Dn"
 #endif
-	.section	.note.GNU-stack,"",@progbits
+	.section	.note.GNU-stack,"",%progbits
