diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared98/locale_facets.cc gcc-6.3.1-20170216/libstdc++-v3/src/nonshared98/locale_facets.cc
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared98/locale_facets.cc	2017-07-06 16:24:59.002668108 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared98/locale_facets.cc	2017-07-06 16:25:38.332524176 +0200
@@ -22,7 +22,7 @@
 
 #define _GLIBCXX_NONSHARED_CXX11
 #include "../c++98/locale_facets.cc"
-#if defined(__i386__) || (defined(__powerpc__) && !defined(__powerpc64__))
+#if defined(__i386__) || defined(__arm__) || (defined(__powerpc__) && !defined(__powerpc64__))
 asm (".hidden _ZSt22__verify_grouping_implPKcjS0_j");
 #else
 asm (".hidden _ZSt22__verify_grouping_implPKcmS0_m");
