--- libstdc++-v3/testsuite/22_locale/locale/cons/40184.cc.jj	2014-04-11 14:52:06.733884798 +0200
+++ libstdc++-v3/testsuite/22_locale/locale/cons/40184.cc	2014-05-10 12:29:51.779887325 +0200
@@ -1,4 +1,6 @@
 // { dg-require-namedlocale "ja_JP.eucjp" }
+// pr40184 fix not backported to base (locales with broken facets)
+// { dg-do run { xfail *-*-* } }
 
 // Copyright (C) 2009-2014 Free Software Foundation, Inc.
 //
--- libstdc++-v3/testsuite/22_locale/time_get/get_date/char/26701.cc.jj	2014-04-11 14:52:06.680885075 +0200
+++ libstdc++-v3/testsuite/22_locale/time_get/get_date/char/26701.cc	2014-05-10 12:29:51.792886917 +0200
@@ -1,4 +1,6 @@
 // { dg-require-namedlocale "en_GB" }
+//  pr26701 fix not backported to base (parsing of en_GB date)
+// { dg-do run { xfail *-*-* } }
 
 // 2010-01-06  Paolo Carlini  <paolo.carlini@oracle.com>
 
--- libstdc++-v3/testsuite/22_locale/time_get/get_date/wchar_t/26701.cc.jj	2014-04-11 14:52:06.674885106 +0200
+++ libstdc++-v3/testsuite/22_locale/time_get/get_date/wchar_t/26701.cc	2014-05-10 12:29:51.792886917 +0200
@@ -1,4 +1,6 @@
 // { dg-require-namedlocale "en_GB" }
+//  pr26701 fix not backported to base (parsing of en_GB date)
+// { dg-do run { xfail *-*-* } }
 
 // 2010-01-06  Paolo Carlini  <paolo.carlini@oracle.com>
 
--- libstdc++-v3/testsuite/22_locale/time_get/get_weekday/char/38081-1.cc.jj	2014-04-11 14:52:06.645885257 +0200
+++ libstdc++-v3/testsuite/22_locale/time_get/get_weekday/char/38081-1.cc	2014-05-10 12:29:51.802886712 +0200
@@ -1,4 +1,6 @@
 // { dg-require-namedlocale "ru_RU.ISO-8859-5" }
+//  pr38081 fix not backported to base (parsing weekday names)
+// { dg-do run { xfail *-*-* } }
 
 // Copyright (C) 2010-2014 Free Software Foundation, Inc.
 //
--- libstdc++-v3/testsuite/22_locale/time_get/get_weekday/char/38081-2.cc.jj	2014-04-11 14:52:06.646885252 +0200
+++ libstdc++-v3/testsuite/22_locale/time_get/get_weekday/char/38081-2.cc	2014-05-10 12:29:51.802886712 +0200
@@ -1,4 +1,6 @@
 // { dg-require-namedlocale "ru_RU.UTF8" }
+// pr38081 fix not backported to base (parsing weekday names)
+// { dg-do run { xfail *-*-* } }
 
 // 2010-01-05  Paolo Carlini  <paolo.carlini@oracle.com>
 
--- libstdc++-v3/testsuite/27_io/basic_filebuf/seekoff/char/4.cc.jj	2014-04-11 14:52:04.949894106 +0200
+++ libstdc++-v3/testsuite/27_io/basic_filebuf/seekoff/char/4.cc	2014-05-10 12:29:51.807886709 +0200
@@ -1,4 +1,6 @@
 // { dg-require-fileio "" }
+// pr45628 fix not backported to base (seek 0 shouldn't flush)
+// { dg-do run { xfail *-*-* } }
 
 // Copyright (C) 2010-2014 Free Software Foundation, Inc.
 //
--- libstdc++-v3/testsuite/27_io/basic_filebuf/seekoff/char/45628-1.cc.jj	2014-05-10 12:29:51.808886705 +0200
+++ libstdc++-v3/testsuite/27_io/basic_filebuf/seekoff/char/45628-1.cc	2014-05-10 12:31:51.652261219 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// pr45628 fix not backported to base (seek 0 shouldn't flush)
+
 // Copyright (C) 2010-2014 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
--- libstdc++-v3/testsuite/27_io/basic_filebuf/seekoff/wchar_t/4.cc.jj	2014-04-11 14:52:04.941894148 +0200
+++ libstdc++-v3/testsuite/27_io/basic_filebuf/seekoff/wchar_t/4.cc	2014-05-10 12:29:51.808886705 +0200
@@ -1,4 +1,6 @@
 // { dg-require-fileio "" }
+// pr45628 fix not backported to base (seek 0 should not flush)
+// { dg-do run { xfail *-*-* } }
 
 // Copyright (C) 2010-2014 Free Software Foundation, Inc.
 //
--- libstdc++-v3/testsuite/27_io/basic_istream/extractors_arithmetic/char/13.cc.jj	2014-04-11 14:52:04.771895035 +0200
+++ libstdc++-v3/testsuite/27_io/basic_istream/extractors_arithmetic/char/13.cc	2014-05-10 12:29:51.819886643 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// dr696 fix not backported to base (handling of overflow when reading number)
+
 // 1999-04-12 bkoz
 
 // Copyright (C) 1999-2014 Free Software Foundation, Inc.
--- libstdc++-v3/testsuite/27_io/basic_istream/extractors_arithmetic/char/dr696.cc.jj	2014-04-11 14:52:04.770895040 +0200
+++ libstdc++-v3/testsuite/27_io/basic_istream/extractors_arithmetic/char/dr696.cc	2014-05-10 12:29:51.819886643 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// dr696 fix not backported to base (handling of overflow when reading number)
+
 // 2009-07-15  Paolo Carlini  <paolo.carlini@oracle.com>
 //
 // Copyright (C) 2009-2014 Free Software Foundation, Inc.
--- libstdc++-v3/testsuite/27_io/basic_istream/extractors_arithmetic/wchar_t/13.cc.jj	2014-05-10 12:29:51.829886586 +0200
+++ libstdc++-v3/testsuite/27_io/basic_istream/extractors_arithmetic/wchar_t/13.cc	2014-05-10 12:32:23.160099254 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// dr696 fix not backported to base (handling of overflow when reading number)
+
 // Copyright (C) 2004-2014 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
--- libstdc++-v3/testsuite/27_io/basic_istream/extractors_arithmetic/wchar_t/dr696.cc.jj	2014-04-11 14:52:04.762895082 +0200
+++ libstdc++-v3/testsuite/27_io/basic_istream/extractors_arithmetic/wchar_t/dr696.cc	2014-05-10 12:29:51.829886586 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// dr696 fix not backported to base (handling of overflow when reading number)
+
 // 2009-07-15  Paolo Carlini  <paolo.carlini@oracle.com>
 //
 // Copyright (C) 2009-2014 Free Software Foundation, Inc.
--- libstdc++-v3/testsuite/27_io/basic_istream/seekg/char/26211.cc.jj	2014-04-11 14:52:04.735895223 +0200
+++ libstdc++-v3/testsuite/27_io/basic_istream/seekg/char/26211.cc	2014-05-10 12:29:51.836886549 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// pr26211 fix not backported to base (tellg/seekg should construct a sentry)
+
 // 2010-11-18  Paolo Carlini  <paolo.carlini@oracle.com>
 
 // Copyright (C) 2010-2014 Free Software Foundation, Inc.
--- libstdc++-v3/testsuite/27_io/basic_istream/seekg/wchar_t/26211.cc.jj	2014-04-11 14:52:04.731895243 +0200
+++ libstdc++-v3/testsuite/27_io/basic_istream/seekg/wchar_t/26211.cc	2014-05-10 12:29:51.836886549 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// pr26211 fix not backported to base (tellg/seekg should construct a sentry)
+
 // 2010-11-18  Paolo Carlini  <paolo.carlini@oracle.com>
 
 // Copyright (C) 2010-2014 Free Software Foundation, Inc.
--- libstdc++-v3/testsuite/27_io/basic_istream/tellg/char/26211.cc.jj	2014-04-11 14:52:04.779894993 +0200
+++ libstdc++-v3/testsuite/27_io/basic_istream/tellg/char/26211.cc	2014-05-10 12:29:51.836886549 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// pr26211 fix not backported to base (tellg/seekg should construct a sentry)
+
 // 2010-11-18  Paolo Carlini  <paolo.carlini@oracle.com>
 
 // Copyright (C) 2010-2014 Free Software Foundation, Inc.
--- libstdc++-v3/testsuite/27_io/basic_istream/tellg/wchar_t/26211.cc.jj	2014-04-11 14:52:04.775895014 +0200
+++ libstdc++-v3/testsuite/27_io/basic_istream/tellg/wchar_t/26211.cc	2014-05-10 12:29:51.837886543 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// pr26211 fix not backported to base (tellg/seekg should construct a sentry)
+
 // 2010-11-18  Paolo Carlini  <paolo.carlini@oracle.com>
 
 // Copyright (C) 2010-2014 Free Software Foundation, Inc.
--- libstdc++-v3/testsuite/27_io/objects/char/41037.cc.jj	2014-05-10 12:29:51.848886483 +0200
+++ libstdc++-v3/testsuite/27_io/objects/char/41037.cc	2014-05-10 12:30:52.480567953 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// pr41037 fix not backported to base (flags on cerr)
+
 // Copyright (C) 2009-2014 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
--- libstdc++-v3/testsuite/27_io/objects/wchar_t/41037.cc.jj	2014-05-10 12:29:51.856886441 +0200
+++ libstdc++-v3/testsuite/27_io/objects/wchar_t/41037.cc	2014-05-10 12:31:20.274421862 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// pr26211 fix not backported to base (tellg/seekg should construct a sentry)
+
 // Copyright (C) 2009-2014 Free Software Foundation, Inc.
 //
 // This file is part of the GNU ISO C++ Library.  This library is free
--- libstdc++-v3/testsuite/abi/demangle/regression/cw-13.cc.jj	2014-04-11 14:52:05.388891816 +0200
+++ libstdc++-v3/testsuite/abi/demangle/regression/cw-13.cc	2014-05-10 12:29:51.868886376 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// pr42670 fix not backported to base (partial demangling of global ctor fn)
+
 // 2003-02-26  Carlo Wood  <carlo@alinoe.com>
 
 // Copyright (C) 2003-2014 Free Software Foundation, Inc.
--- libstdc++-v3/testsuite/abi/demangle/regression/cw-16.cc.jj	2014-04-11 14:52:05.387891821 +0200
+++ libstdc++-v3/testsuite/abi/demangle/regression/cw-16.cc	2014-05-10 12:29:51.868886376 +0200
@@ -1,3 +1,6 @@
+// { dg-do run { xfail *-*-* } }
+// pr37376 fix not backported to base ((42) + 2 changed to (42)+(2))
+
 // 2003-11-07  Carlo Wood  <carlo@alinoe.com>
 
 // Copyright (C) 2003-2014 Free Software Foundation, Inc.
--- libstdc++-v3/testsuite/libstdc++-abi/abi.exp.jj	2014-04-11 14:52:06.793884485 +0200
+++ libstdc++-v3/testsuite/libstdc++-abi/abi.exp	2014-05-10 12:29:51.868886376 +0200
@@ -61,4 +61,7 @@ remote_download "target" "current_symbol
 set result [${tool}_load "./abi_check" \
 		[list "--check-verbose" "current_symbols.txt" \
 		     "baseline_symbols.txt"]]
+# This is an artifact of running the current testsuite against the old
+# system libstdc++
+setup_xfail *-*-*
 [lindex $result 0] "libstdc++-abi/abi_check"
