diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/cow-shim_facets.cc gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/cow-shim_facets.cc
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/cow-shim_facets.cc	2017-07-03 20:44:15.375137096 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/cow-shim_facets.cc	2017-07-03 21:51:43.270512105 +0200
@@ -168,7 +168,7 @@
 asm (".hidden _ZTVNSt13__facet_shims12_GLOBAL__N_115moneypunct_shimIwLb1EEE");
 asm (".hidden _ZNSt13__facet_shims21__numpunct_fill_cacheIcEEvSt17integral_constantIbLb0EEPKNSt6locale5facetEPSt16__numpunct_cacheIT_E");
 asm (".hidden _ZNSt13__facet_shims21__numpunct_fill_cacheIwEEvSt17integral_constantIbLb0EEPKNSt6locale5facetEPSt16__numpunct_cacheIT_E");
-#if defined(__i386__) || (defined(__powerpc__) && !defined(__powerpc64__))
+#if defined(__i386__) || defined(__arm__) || (defined(__powerpc__) && !defined(__powerpc64__))
 asm (".hidden _ZNSt13__facet_shims14__messages_getIcEEvSt17integral_constantIbLb0EEPKNSt6locale5facetERNS_12__any_stringEiiiPKT_j");
 asm (".hidden _ZNSt13__facet_shims14__messages_getIwEEvSt17integral_constantIbLb0EEPKNSt6locale5facetERNS_12__any_stringEiiiPKT_j");
 asm (".hidden _ZNSt13__facet_shims15__messages_openIcEEiSt17integral_constantIbLb0EEPKNSt6locale5facetEPKcjRKS3_");
diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/cow-stdexcept.cc gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/cow-stdexcept.cc
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/cow-stdexcept.cc	2017-07-03 20:44:15.425136949 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/cow-stdexcept.cc	2017-07-03 21:51:48.600488066 +0200
@@ -35,7 +35,7 @@
 asm (".hidden _ZNSt12__cow_stringaSERKS_");
 asm (".hidden _ZNSt12__sso_stringC1ERKSs");
 asm (".hidden _ZNSt12__sso_stringC2ERKSs");
-#if defined(__i386__) || (defined(__powerpc__) && !defined(__powerpc64__))
+#if defined(__i386__) || defined(__arm__) || (defined(__powerpc__) && !defined(__powerpc64__))
 asm (".hidden _ZNSt12__cow_stringC1EPKcj");
 asm (".hidden _ZNSt12__cow_stringC2EPKcj");
 #else
diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/cxx11-shim_facets.cc gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/cxx11-shim_facets.cc
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/cxx11-shim_facets.cc	2017-07-03 20:44:15.470136818 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/cxx11-shim_facets.cc	2017-07-03 21:51:54.210462764 +0200
@@ -168,7 +168,7 @@
 asm (".hidden _ZTVNSt13__facet_shims12_GLOBAL__N_115moneypunct_shimIcLb1EEE");
 asm (".hidden _ZTVNSt13__facet_shims12_GLOBAL__N_115moneypunct_shimIwLb0EEE");
 asm (".hidden _ZTVNSt13__facet_shims12_GLOBAL__N_115moneypunct_shimIwLb1EEE");
-#if defined(__i386__) || (defined(__powerpc__) && !defined(__powerpc64__))
+#if defined(__i386__) || defined(__arm__) || (defined(__powerpc__) && !defined(__powerpc64__))
 asm (".hidden _ZNSt13__facet_shims14__messages_getIcEEvSt17integral_constantIbLb1EEPKNSt6locale5facetERNS_12__any_stringEiiiPKT_j");
 asm (".hidden _ZNSt13__facet_shims14__messages_getIwEEvSt17integral_constantIbLb1EEPKNSt6locale5facetERNS_12__any_stringEiiiPKT_j");
 asm (".hidden _ZNSt13__facet_shims15__messages_openIcEEiSt17integral_constantIbLb1EEPKNSt6locale5facetEPKcjRKS3_");
diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/cxx11-wlocale-inst.cc gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/cxx11-wlocale-inst.cc
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/cxx11-wlocale-inst.cc	2017-07-03 20:44:15.520136671 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/cxx11-wlocale-inst.cc	2017-07-03 21:52:00.745433290 +0200
@@ -38,7 +38,7 @@
 asm (".hidden _ZNKSt17__gnu_cxx_ldbl1287num_getIwSt19istreambuf_iteratorIwSt11char_traitsIwEEE14_M_extract_intB5cxx11IxEES4_S4_S4_RSt8ios_baseRSt12_Ios_IostateRT_");
 asm (".hidden _ZNKSt17__gnu_cxx_ldbl1287num_getIwSt19istreambuf_iteratorIwSt11char_traitsIwEEE14_M_extract_intB5cxx11IyEES4_S4_S4_RSt8ios_baseRSt12_Ios_IostateRT_");
 #endif
-#if defined(__i386__) || (defined(__powerpc__) && !defined(__powerpc64__))
+#if defined(__i386__) || defined(__arm__) || (defined(__powerpc__) && !defined(__powerpc64__))
 asm (".hidden _ZNSt15time_put_bynameIwSt19ostreambuf_iteratorIwSt11char_traitsIwEEEC1ERKNSt7__cxx1112basic_stringIcS1_IcESaIcEEEj");
 asm (".hidden _ZNSt15time_put_bynameIwSt19ostreambuf_iteratorIwSt11char_traitsIwEEEC2ERKNSt7__cxx1112basic_stringIcS1_IcESaIcEEEj");
 #else
diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/locale-inst.cc gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/locale-inst.cc
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/locale-inst.cc	2017-07-03 20:44:15.565136539 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/locale-inst.cc	2017-07-03 21:52:07.815401403 +0200
@@ -78,14 +78,14 @@
 asm (".hidden _ZTSSt9time_base");
 #ifdef C_is_char
 asm (".hidden _ZNSt19istreambuf_iteratorIcSt11char_traitsIcEEppEv");
-#if defined(__i386__) || (defined(__powerpc__) && !defined(__powerpc64__))
+#if defined(__i386__) || defined(__arm__) || (defined(__powerpc__) && !defined(__powerpc64__))
 asm (".hidden _ZSt14__add_groupingIcEPT_S1_S0_PKcjPKS0_S5_");
 #else
 asm (".hidden _ZSt14__add_groupingIcEPT_S1_S0_PKcmPKS0_S5_");
 #endif
 #else
 asm (".hidden _ZNSt19istreambuf_iteratorIwSt11char_traitsIwEEppEv");
-#if defined(__i386__) || (defined(__powerpc__) && !defined(__powerpc64__))
+#if defined(__i386__) || defined(__arm__) || (defined(__powerpc__) && !defined(__powerpc64__))
 asm (".hidden _ZSt14__add_groupingIwEPT_S1_S0_PKcjPKS0_S5_");
 #else
 asm (".hidden _ZSt14__add_groupingIwEPT_S1_S0_PKcmPKS0_S5_");
diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/snprintf_lite.cc gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/snprintf_lite.cc
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/snprintf_lite.cc	2017-07-03 20:44:15.615136393 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/snprintf_lite.cc	2017-07-03 22:00:30.333166823 +0200
@@ -30,6 +30,9 @@
 #elif defined(__powerpc__)
 asm (".hidden _ZN9__gnu_cxx15__concat_size_tEPcjj");
 asm (".hidden _ZN9__gnu_cxx15__snprintf_liteEPcjPKcP13__va_list_tag");
+#elif defined(__arm__)
+asm (".hidden _ZN9__gnu_cxx15__concat_size_tEPcjj");
+asm (".hidden _ZN9__gnu_cxx15__snprintf_liteEPcjPKcSt9__va_list");
 #elif defined(__aarch64__)
 asm (".hidden _ZN9__gnu_cxx15__concat_size_tEPcmm");
 asm (".hidden _ZN9__gnu_cxx15__snprintf_liteEPcmPKcSt9__va_list");
diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/system_error44.cc gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/system_error44.cc
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/system_error44.cc	2017-07-03 20:44:15.660136261 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/system_error44.cc	2017-07-03 21:52:30.665298348 +0200
@@ -37,7 +37,7 @@
 asm (".hidden _ZNSt12__sso_stringaSEOS_");
 asm (".hidden _ZNSt12__sso_stringaSERKS_");
 asm (".hidden _ZTSNSt3_V214error_categoryE");
-#if defined(__i386__) || (defined(__powerpc__) && !defined(__powerpc64__))
+#if defined(__i386__) || defined(__arm__) || (defined(__powerpc__) && !defined(__powerpc64__))
 asm (".hidden _ZNSt12__sso_stringC1EPKcj");
 asm (".hidden _ZNSt12__sso_stringC2EPKcj");
 #else
diff -ru gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/system_error48.cc gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/system_error48.cc
--- gcc-6.3.1-20170216-orig/libstdc++-v3/src/nonshared11/system_error48.cc	2017-07-03 20:44:15.710136115 +0200
+++ gcc-6.3.1-20170216/libstdc++-v3/src/nonshared11/system_error48.cc	2017-07-03 21:52:39.995256269 +0200
@@ -37,7 +37,7 @@
 asm (".hidden _ZNSt12__sso_stringaSEOS_");
 asm (".hidden _ZNSt12__sso_stringaSERKS_");
 asm (".hidden _ZTSNSt3_V214error_categoryE");
-#if defined(__i386__) || (defined(__powerpc__) && !defined(__powerpc64__))
+#if defined(__i386__) || defined(__arm__) || (defined(__powerpc__) && !defined(__powerpc64__))
 asm (".hidden _ZNSt12__sso_stringC1EPKcj");
 asm (".hidden _ZNSt12__sso_stringC2EPKcj");  
 #else
