From 23d7f34af3e6974eea43abe1ca9b3054822014cf Mon Sep 17 00:00:00 2001
From: Jan Chaloupka <jchaloup@redhat.com>
Date: Mon, 10 Aug 2015 09:55:49 +0200
Subject: [PATCH] Change etcd server port

---
 contrib/init/systemd/environ/apiserver | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/contrib/init/systemd/environ/apiserver b/contrib/init/systemd/environ/apiserver
index 223b992..ddb5974 100644
--- a/contrib/init/systemd/environ/apiserver
+++ b/contrib/init/systemd/environ/apiserver
@@ -14,7 +14,7 @@ KUBE_API_ADDRESS="--insecure-bind-address=127.0.0.1"
 # KUBELET_PORT="--kubelet-port=10250"
 
 # Comma separated list of nodes in the etcd cluster
-KUBE_ETCD_SERVERS="--etcd-servers=http://127.0.0.1:4001"
+KUBE_ETCD_SERVERS="--etcd-servers=http://127.0.0.1:2379"
 
 # Address range to use for services
 KUBE_SERVICE_ADDRESSES="--service-cluster-ip-range=10.254.0.0/16"
-- 
1.9.3

