From 916a2fa6cca162faa2e62d121c128901662d2e48 Mon Sep 17 00:00:00 2001
From: Jan Chaloupka <jchaloup@redhat.com>
Date: Mon, 6 Feb 2017 22:49:21 +0100
Subject: [PATCH] hack/test-cmd.sh

---
 hack/lib/init.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hack/lib/init.sh b/hack/lib/init.sh
index ae50779..0a601ac 100644
--- a/hack/lib/init.sh
+++ b/hack/lib/init.sh
@@ -42,7 +42,7 @@ source "${KUBE_ROOT}/hack/lib/version.sh"
 source "${KUBE_ROOT}/hack/lib/golang.sh"
 source "${KUBE_ROOT}/hack/lib/etcd.sh"
 
-KUBE_OUTPUT_HOSTBIN="${KUBE_OUTPUT_BINPATH}/$(kube::util::host_platform)"
+KUBE_OUTPUT_HOSTBIN="/usr/bin"
 
 # list of all available group versions.  This should be used when generated code
 # or when starting an API server that you want to have everything.
-- 
2.7.4

