diff -urNp coreutils-8.22-orig/tests/df/df-symlink.sh coreutils-8.22/tests/df/df-symlink.sh
--- coreutils-8.22-orig/tests/df/df-symlink.sh	2013-12-04 15:48:30.000000000 +0100
+++ coreutils-8.22/tests/df/df-symlink.sh	2013-12-14 18:20:15.822594995 +0100
@@ -18,6 +18,9 @@
 
 . "${srcdir=.}/tests/init.sh"; path_prepend_ ./src
 print_ver_ df
+#df doesn't work correctly on symlinks when on LVM/LUKS filesystem, therefore
+#marking expensive_ to disable by default
+expensive_
 
 disk=$(df --out=source '.' | tail -n1) ||
   skip_ "cannot determine '.' file system"
