diff -urNp coreutils-8.22-orig/src/cp.c coreutils-8.22/src/cp.c
--- coreutils-8.22-orig/src/cp.c	2015-06-11 15:58:04.230858212 +0200
+++ coreutils-8.22/src/cp.c	2015-06-11 15:59:13.191396755 +0200
@@ -233,8 +233,10 @@ Copy SOURCE to DEST, or multiple SOURCE(
   -x, --one-file-system        stay on this file system\n\
 "), stdout);
       fputs (_("\
-  -Z, --context[=CTX]          set SELinux security context of destination\n\
-                                 file to default type, or to CTX if specified\n\
+  -Z                           set SELinux security context of destination\n\
+                                 file to default type\n\
+      --context[=CTX]          like -Z, or if CTX is specified then set the\n\
+                                 SELinux or SMACK security context to CTX\n\
 "), stdout);
       fputs (HELP_OPTION_DESCRIPTION, stdout);
       fputs (VERSION_OPTION_DESCRIPTION, stdout);
diff -urNp coreutils-8.22-orig/src/install.c coreutils-8.22/src/install.c
--- coreutils-8.22-orig/src/install.c	2015-06-11 15:58:04.230858212 +0200
+++ coreutils-8.22/src/install.c	2015-06-11 16:00:16.754893027 +0200
@@ -647,8 +647,10 @@ In the 4th form, create all components o
 "), stdout);
       fputs (_("\
   -P, --preserve-context  preserve SELinux security context (-P deprecated)\n\
-  -Z, --context[=CTX]     set SELinux security context of destination file to\n\
-                            default type, or to CTX if specified\n\
+  -Z                      set SELinux security context of destination\n\
+                            file to default type\n\
+      --context[=CTX]     like -Z, or if CTX is specified then set the\n\
+                            SELinux or SMACK security context to CTX\n\
 "), stdout);
 
       fputs (HELP_OPTION_DESCRIPTION, stdout);
diff -urNp coreutils-8.22-orig/src/mkdir.c coreutils-8.22/src/mkdir.c
--- coreutils-8.22-orig/src/mkdir.c	2013-12-05 01:59:36.000000000 +0100
+++ coreutils-8.22/src/mkdir.c	2015-06-11 16:01:17.209364915 +0200
@@ -66,8 +66,12 @@ Create the DIRECTORY(ies), if they do no
   -m, --mode=MODE   set file mode (as in chmod), not a=rwx - umask\n\
   -p, --parents     no error if existing, make parent directories as needed\n\
   -v, --verbose     print a message for each created directory\n\
-  -Z, --context[=CTX]  set the SELinux security context of each created\n\
-                         directory to default type or to CTX if specified\n\
+"), stdout);
+      fputs (_("\
+  -Z                   set SELinux security context of each created directory\n\
+                         to the default type\n\
+      --context[=CTX]  like -Z, or if CTX is specified then set the SELinux\n\
+                         or SMACK security context to CTX\n\
 "), stdout);
       fputs (HELP_OPTION_DESCRIPTION, stdout);
       fputs (VERSION_OPTION_DESCRIPTION, stdout);
diff -urNp coreutils-8.22-orig/src/mkfifo.c coreutils-8.22/src/mkfifo.c
--- coreutils-8.22-orig/src/mkfifo.c	2013-12-05 00:43:05.000000000 +0100
+++ coreutils-8.22/src/mkfifo.c	2015-06-11 16:02:03.389725315 +0200
@@ -61,8 +61,9 @@ Create named pipes (FIFOs) with the give
   -m, --mode=MODE    set file permission bits to MODE, not a=rw - umask\n\
 "), stdout);
       fputs (_("\
-  -Z, --context[=CTX]  set the SELinux security context of each NAME to\n\
-                         default type, or CTX if specified\n\
+  -Z                   set the SELinux security context to default type\n\
+      --context[=CTX]  like -Z, or if CTX is specified then set the SELinux\n\
+                         or SMACK security context to CTX\n\
 "), stdout);
       fputs (HELP_OPTION_DESCRIPTION, stdout);
       fputs (VERSION_OPTION_DESCRIPTION, stdout);
diff -urNp coreutils-8.22-orig/src/mknod.c coreutils-8.22/src/mknod.c
--- coreutils-8.22-orig/src/mknod.c	2013-12-05 00:43:05.000000000 +0100
+++ coreutils-8.22/src/mknod.c	2015-06-11 16:02:36.300982160 +0200
@@ -63,8 +63,9 @@ Create the special file NAME of the give
   -m, --mode=MODE    set file permission bits to MODE, not a=rw - umask\n\
 "), stdout);
       fputs (_("\
-  -Z, --context[=CTX]  set the SELinux security context of NAME to\n\
-                         default type, or to CTX if specified\n\
+  -Z                   set the SELinux security context to default type\n\
+      --context[=CTX]  like -Z, or if CTX is specified then set the SELinux\n\
+                         or SMACK security context to CTX\n\
 "), stdout);
       fputs (HELP_OPTION_DESCRIPTION, stdout);
       fputs (VERSION_OPTION_DESCRIPTION, stdout);
