diff -urNp coreutils-8.22-orig/src/fs-is-local.h coreutils-8.22/src/fs-is-local.h
--- coreutils-8.22-orig/src/fs-is-local.h	2016-06-24 10:59:08.545965484 +0200
+++ coreutils-8.22/src/fs-is-local.h	2016-06-24 11:08:47.435944633 +0200
@@ -6,6 +6,7 @@ is_local_fs_type (unsigned long int magi
 {
   switch (magic)
     {
+      case S_MAGIC_ACFS: return 0;
       case S_MAGIC_ADFS: return 1;
       case S_MAGIC_AFFS: return 1;
       case S_MAGIC_AFS: return 0;
@@ -15,13 +16,16 @@ is_local_fs_type (unsigned long int magi
       case S_MAGIC_BEFS: return 1;
       case S_MAGIC_BDEVFS: return 1;
       case S_MAGIC_BFS: return 1;
+      case S_MAGIC_BPF_FS: return 1;
       case S_MAGIC_BINFMTFS: return 1;
       case S_MAGIC_BTRFS: return 1;
+      case S_MAGIC_BTRFS_TEST: return 1;
       case S_MAGIC_CEPH: return 0;
       case S_MAGIC_CGROUP: return 1;
       case S_MAGIC_CIFS: return 0;
       case S_MAGIC_CODA: return 0;
       case S_MAGIC_COH: return 1;
+      case S_MAGIC_CONFIGFS: return 1;
       case S_MAGIC_CRAMFS: return 1;
       case S_MAGIC_CRAMFS_WEND: return 1;
       case S_MAGIC_DEBUGFS: return 1;
@@ -43,10 +47,13 @@ is_local_fs_type (unsigned long int magi
       case S_MAGIC_GFS: return 0;
       case S_MAGIC_GPFS: return 0;
       case S_MAGIC_HFS: return 1;
+      case S_MAGIC_HFS_PLUS: return 1;
+      case S_MAGIC_HFS_X: return 1;
       case S_MAGIC_HOSTFS: return 1;
       case S_MAGIC_HPFS: return 1;
       case S_MAGIC_HUGETLBFS: return 1;
       case S_MAGIC_MTD_INODE_FS: return 1;
+      case S_MAGIC_IBRIX: return 0;
       case S_MAGIC_INOTIFYFS: return 1;
       case S_MAGIC_ISOFS: return 1;
       case S_MAGIC_ISOFS_R_WIN: return 1;
@@ -55,7 +62,9 @@ is_local_fs_type (unsigned long int magi
       case S_MAGIC_JFFS2: return 1;
       case S_MAGIC_JFS: return 1;
       case S_MAGIC_KAFS: return 0;
+      case S_MAGIC_LOGFS: return 1;
       case S_MAGIC_LUSTRE: return 0;
+      case S_MAGIC_M1FS: return 1;
       case S_MAGIC_MINIX: return 1;
       case S_MAGIC_MINIX_30: return 1;
       case S_MAGIC_MINIX_V2: return 1;
@@ -67,11 +75,14 @@ is_local_fs_type (unsigned long int magi
       case S_MAGIC_NFS: return 0;
       case S_MAGIC_NFSD: return 0;
       case S_MAGIC_NILFS: return 1;
+      case S_MAGIC_NSFS: return 1;
       case S_MAGIC_NTFS: return 1;
       case S_MAGIC_OPENPROM: return 1;
       case S_MAGIC_OCFS2: return 0;
+      case S_MAGIC_OVERLAYFS: return 0;
       case S_MAGIC_PANFS: return 0;
       case S_MAGIC_PIPEFS: return 0;
+      case S_MAGIC_PRL_FS: return 0;
       case S_MAGIC_PROC: return 1;
       case S_MAGIC_PSTOREFS: return 1;
       case S_MAGIC_QNX4: return 1;
@@ -91,6 +101,7 @@ is_local_fs_type (unsigned long int magi
       case S_MAGIC_SYSV2: return 1;
       case S_MAGIC_SYSV4: return 1;
       case S_MAGIC_TMPFS: return 1;
+      case S_MAGIC_TRACEFS: return 1;
       case S_MAGIC_UBIFS: return 1;
       case S_MAGIC_UDF: return 1;
       case S_MAGIC_UFS: return 1;
diff -urNp coreutils-8.22-orig/src/stat.c coreutils-8.22/src/stat.c
--- coreutils-8.22-orig/src/stat.c	2013-12-13 15:12:46.000000000 +0100
+++ coreutils-8.22/src/stat.c	2016-06-24 11:16:37.853990792 +0200
@@ -311,13 +311,13 @@ human_fstype (STRUCT_STATVFS const *stat
       return "fusectl";
     case S_MAGIC_FUTEXFS: /* 0x0BAD1DEA local */
       return "futexfs";
-    case S_MAGIC_GFS: /* 0x1161970 remote */
+    case S_MAGIC_GFS: /* 0x01161970 remote */
       return "gfs/gfs2";
     case S_MAGIC_GPFS: /* 0x47504653 remote */
       return "gpfs";
     case S_MAGIC_HFS: /* 0x4244 local */
       return "hfs";
-    case S_MAGIC_HOSTFS: /* 0xC0FFEE local */
+    case S_MAGIC_HOSTFS: /* 0x00C0FFEE local */
       return "hostfs";
     case S_MAGIC_HPFS: /* 0xF995E849 local */
       return "hpfs";
@@ -369,7 +369,7 @@ human_fstype (STRUCT_STATVFS const *stat
       return "ntfs";
     case S_MAGIC_OPENPROM: /* 0x9FA1 local */
       return "openprom";
-    case S_MAGIC_OCFS2: /* 0x7461636f remote */
+    case S_MAGIC_OCFS2: /* 0x7461636F remote */
       return "ocfs2";
     case S_MAGIC_PANFS: /* 0xAAD7AAEA remote */
       return "panfs";
@@ -430,7 +430,9 @@ human_fstype (STRUCT_STATVFS const *stat
       return "v9fs";
     case S_MAGIC_VMHGFS: /* 0xBACBACBC remote */
       return "vmhgfs";
-    case S_MAGIC_VXFS: /* 0xA501FCF5 local */
+    case S_MAGIC_VXFS: /* 0xA501FCF5 remote */
+       /* Veritas File System can run in single instance or clustered mode,
+          so mark as remote to cater for the latter case.  */
       return "vxfs";
     case S_MAGIC_VZFS: /* 0x565A4653 local */
       return "vzfs";
@@ -444,6 +446,37 @@ human_fstype (STRUCT_STATVFS const *stat
       return "xia";
     case S_MAGIC_ZFS: /* 0x2FC12FC1 local */
       return "zfs";
+/*  Refresh from coreutils-8.25 bellow  */
+    case S_MAGIC_ACFS: /* 0x61636673 remote */
+      return "acfs";
+    case S_MAGIC_BPF_FS: /* 0xCAFE4A11 local */
+      return "bpf_fs";
+    case S_MAGIC_BTRFS_TEST: /* 0x73727279 local */
+      return "btrfs_test";
+    case S_MAGIC_CONFIGFS: /* 0x62656570 local */
+      return "configfs";
+    case S_MAGIC_HFS_PLUS: /* 0x482B local */
+      return "hfs+";
+    case S_MAGIC_HFS_X: /* 0x4858 local */
+      return "hfsx";
+    case S_MAGIC_IBRIX: /* 0x013111A8 remote */
+      return "ibrix";
+    case S_MAGIC_LOGFS: /* 0xC97E8168 local */
+      return "logfs";
+    case S_MAGIC_M1FS: /* 0x5346314D local */
+      return "m1fs";
+    case S_MAGIC_NSFS: /* 0x6E736673 local */
+      return "nsfs";
+    case S_MAGIC_OVERLAYFS: /* 0x794C7630 remote */
+       /* This may overlay remote file systems.
+          Also there have been issues reported with inotify and overlayfs,
+          so mark as "remote" so that polling is used.  */
+      return "overlayfs";
+    case S_MAGIC_PRL_FS: /* 0x7C7C6673 remote */
+      return "prl_fs";
+    case S_MAGIC_TRACEFS: /* 0x74726163 local */
+      return "tracefs";
+
 
 # elif __GNU__
     case FSTYPE_UFS:
diff -urNp coreutils-8.22-orig/src/tail.c coreutils-8.22/src/tail.c
--- coreutils-8.22-orig/src/tail.c	2013-12-09 14:40:46.000000000 +0100
+++ coreutils-8.22/src/tail.c	2016-06-24 11:03:10.268044584 +0200
@@ -898,13 +898,15 @@ fremote (int fd, const char *name)
         case 0:
           break;
         case -1:
+#if 0
           {
             unsigned long int fs_type = buf.f_type;
             error (0, 0, _("unrecognized file system type 0x%08lx for %s. "
                            "please report this to %s. reverting to polling"),
                    fs_type, quote (name), PACKAGE_BUGREPORT);
             /* Treat as "remote", so caller polls.  */
           }
+#endif
           break;
         case 1:
           remote = false;
