diff -urNp coreutils-8.22-orig/tests/cp/cp-a-selinux.sh coreutils-8.22/tests/cp/cp-a-selinux.sh
--- coreutils-8.22-orig/tests/cp/cp-a-selinux.sh	2016-06-24 11:44:21.909303666 +0200
+++ coreutils-8.22/tests/cp/cp-a-selinux.sh	2016-06-24 11:47:28.473908346 +0200
@@ -73,7 +73,8 @@ old_type_d=$(get_selinux_type c_d)
 # and get the adjusted type for comparison
 cp -a c Z1 || fail=1
 cp -a c_d Z1_d || fail=1
-if restorecon Z1 Z1_d 2>/dev/null; then
+if restorecon Z1 Z1_d 2>restorecon.err \
+   && compare /dev/null restorecon.err; then
   new_type_f=$(get_selinux_type Z1)
   new_type_d=$(get_selinux_type Z1_d)
 
diff -urNp coreutils-8.22-orig/tests/local.mk coreutils-8.22/tests/local.mk
--- coreutils-8.22-orig/tests/local.mk	2016-06-24 11:44:21.920303760 +0200
+++ coreutils-8.22/tests/local.mk	2016-06-24 11:49:31.388965563 +0200
@@ -117,7 +117,6 @@ all_root_tests =				\
   tests/df/problematic-chars.sh			\
   tests/du/bind-mount-dir-cycle.sh		\
   tests/du/bind-mount-dir-cycle-v2.sh		\
-  tests/id/setgid.sh				\
   tests/install/install-C-root.sh		\
   tests/ls/capability.sh			\
   tests/ls/nameless-uid.sh			\
@@ -515,7 +514,6 @@ all_tests =					\
   tests/id/no-context.sh			\
   tests/id/context.sh				\
   tests/id/uid.sh				\
-  tests/id/setgid.sh				\
   tests/id/zero.sh				\
   tests/install/basic-1.sh			\
   tests/install/create-leading.sh		\
