diff -urNp coreutils-8.22-orig/src/fs-is-local.h coreutils-8.22/src/fs-is-local.h
--- coreutils-8.22-orig/src/fs-is-local.h	2013-12-13 15:17:10.000000000 +0100
+++ coreutils-8.22/src/fs-is-local.h	2015-06-04 11:06:34.161072669 +0200
@@ -98,7 +98,7 @@ is_local_fs_type (unsigned long int magi
       case S_MAGIC_USBDEVFS: return 1;
       case S_MAGIC_V9FS: return 1;
       case S_MAGIC_VMHGFS: return 0;
-      case S_MAGIC_VXFS: return 1;
+      case S_MAGIC_VXFS: return 0;
       case S_MAGIC_VZFS: return 1;
       case S_MAGIC_XENFS: return 1;
       case S_MAGIC_XENIX: return 1;
