--- jdk8/jdk/src/share/lib/security/java.security-linux.orig
+++ jdk8/jdk/src/share/lib/security/java.security-linux
@@ -223,7 +223,9 @@
                jdk.nashorn.internal.,\
                jdk.nashorn.tools.,\
                jdk.xml.internal.,\
-               com.sun.activation.registries.
+               com.sun.activation.registries.,\
+               org.GNOME.Accessibility.,\
+               org.GNOME.Bonobo.
 
 #
 # List of comma-separated packages that start with or equal this string
@@ -273,7 +275,9 @@
                    jdk.nashorn.internal.,\
                    jdk.nashorn.tools.,\
                    jdk.xml.internal.,\
-                   com.sun.activation.registries.
+                   com.sun.activation.registries.,\
+                   org.GNOME.Accessibility.,\
+                   org.GNOME.Bonobo.
 
 #
 # Determines whether this properties file can be appended to
