From 3246a236f7d959455137cc0e56a6a9d78d3ba5ec Mon Sep 17 00:00:00 2001
From: Klaus Wenninger <klaus.wenninger@aon.at>
Date: Tue, 21 Mar 2017 02:54:50 +0100
Subject: [PATCH] Fix: pacemaker_remoted: Allow compilation with glib older
 2.36

---
 lrmd/main.c | 8 ++------
 1 file changed, 2 insertions(+), 6 deletions(-)

diff --git a/lrmd/main.c b/lrmd/main.c
index a3aa08f..0e1223c 100644
--- a/lrmd/main.c
+++ b/lrmd/main.c
@@ -382,12 +382,8 @@ void handle_shutdown_nack()
     if (shutting_down) {
         crm_info("Received shutdown nack");
         if (shutdown_ack_timer > 0) {
-            GSource *timer =
-                g_main_context_find_source_by_id(NULL, shutdown_ack_timer);
-
-            if (timer != NULL) {
-                g_source_set_ready_time(timer, 0);
-            }
+            g_source_remove(shutdown_ack_timer);
+            shutdown_ack_timer = g_timeout_add(0, lrmd_exit, NULL);
         }
         return;
     }
-- 
1.8.3.1

