From c757d842b4fbc9a056f1ebaf212a64e9196b2b2d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ferenc=20W=C3=A1gner?= <wferi@debian.org>
Date: Sat, 5 Nov 2016 03:25:32 +0100
Subject: [PATCH] Fix spelling: Negligable -> Negligible

---
 crmd/throttle.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/crmd/throttle.c b/crmd/throttle.c
index ae371c6..ce330fe 100644
--- a/crmd/throttle.c
+++ b/crmd/throttle.c
@@ -377,7 +377,7 @@ throttle_handle_load(float load, const char *desc, int cores)
         return throttle_low;
     }
 
-    crm_trace("Negligable %s detected: %f", desc, adjusted_load);
+    crm_trace("Negligible %s detected: %f", desc, adjusted_load);
     return throttle_none;
 }
 
@@ -433,7 +433,7 @@ throttle_mode(void)
             mode |= throttle_low;
 
         } else {
-            crm_trace("Negligable %s detected: %f", desc, load);
+            crm_trace("Negligible %s detected: %f", desc, load);
         }
     }
 
-- 
1.8.3.1

