From 4ad87e494d56e18bcaafa058554573f890517eed Mon Sep 17 00:00:00 2001
From: Klaus Wenninger <klaus.wenninger@aon.at>
Date: Fri, 21 Jul 2017 17:57:48 +0200
Subject: [PATCH] Fix: stonith-ng: make fencing-device reappear properly after
 reenabling

---
 fencing/main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fencing/main.c b/fencing/main.c
index 360bc25..f58ccc7 100644
--- a/fencing/main.c
+++ b/fencing/main.c
@@ -781,7 +781,7 @@ update_cib_stonith_devices_v2(const char *event, xmlNode * msg)
             }
             free(mutable);
 
-        } else if(strstr(xpath, "/"XML_CIB_TAG_RESOURCES)) {
+        } else if(strstr(xpath, XML_CIB_TAG_RESOURCES)) {
             shortpath = strrchr(xpath, '/'); CRM_ASSERT(shortpath);
             reason = crm_strdup_printf("%s %s", op, shortpath+1);
             needs_update = TRUE;
-- 
1.8.3.1

