From e7027e9d303be5e3f9531c0cb0ef8af914f2adda Mon Sep 17 00:00:00 2001
From: Klaus Wenninger <klaus.wenninger@aon.at>
Date: Tue, 11 Jul 2017 16:35:33 +0200
Subject: [PATCH] fix: stonith-ng: avoid double-free of pending-ops in
 free_device

---
 fencing/commands.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/fencing/commands.c b/fencing/commands.c
index b336d3b..57a1a4e 100644
--- a/fencing/commands.c
+++ b/fencing/commands.c
@@ -509,7 +509,6 @@ free_device(gpointer data)
 
         crm_warn("Removal of device '%s' purged operation %s", device->id, cmd->action);
         cmd->done_cb(0, -ENODEV, NULL, cmd);
-        free_async_command(cmd);
     }
     g_list_free(device->pending_ops);
 
-- 
1.8.3.1

