From 84e1eb06f3434edc20d87d3d3b5d09c755eeee69 Mon Sep 17 00:00:00 2001
From: Andrew Beekhof <andrew@beekhof.net>
Date: Fri, 5 May 2017 18:51:49 +1000
Subject: [PATCH] container: The location of the docker resource is the more
 interesting one

---
 lib/pengine/container.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/lib/pengine/container.c b/lib/pengine/container.c
index 1280227..043f129 100644
--- a/lib/pengine/container.c
+++ b/lib/pengine/container.c
@@ -883,9 +883,9 @@ tuple_print(container_grouping_t * tuple, const char *pre_text, long options, vo
         offset += snprintf(buffer + offset, LINE_MAX - offset, " (%s)", tuple->ipaddr);
     }
 
-    if(tuple->remote && tuple->remote->running_on != NULL) {
-        node = tuple->remote->running_on->data;
-    } else if (tuple->remote == NULL && rsc->running_on != NULL) {
+    if(tuple->docker && tuple->docker->running_on != NULL) {
+        node = tuple->docker->running_on->data;
+    } else if (tuple->docker == NULL && rsc->running_on != NULL) {
         node = rsc->running_on->data;
     }
     common_print(rsc, pre_text, buffer, node, options, print_data);
-- 
1.8.3.1

