From a82c9726299044f4d9d0cfac6ed78a959db726c2 Mon Sep 17 00:00:00 2001
From: Ken Gaillot <kgaillot@redhat.com>
Date: Mon, 2 Oct 2017 15:51:31 -0500
Subject: [PATCH] Low: libpe_status: avoid memory leak in unfencing digests

---
 lib/pengine/utils.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/lib/pengine/utils.c b/lib/pengine/utils.c
index b58f05a..82b6809 100644
--- a/lib/pengine/utils.c
+++ b/lib/pengine/utils.c
@@ -2157,10 +2157,12 @@ pe_fence_op(node_t * node, const char *op, bool optional, const char *reason, pe
                     digests_secure+digests_secure_offset, max-digests_secure_offset,
                     "%s:%s:%s,", match->id, (const char*)g_hash_table_lookup(match->meta, XML_ATTR_TYPE), data->digest_secure_calc);
             }
-            add_hash_param(stonith_op->meta, strdup(XML_OP_ATTR_DIGESTS_ALL),
-                           digests_all);
-            add_hash_param(stonith_op->meta, strdup(XML_OP_ATTR_DIGESTS_SECURE),
-                           digests_secure);
+            g_hash_table_insert(stonith_op->meta,
+                                strdup(XML_OP_ATTR_DIGESTS_ALL),
+                                digests_all);
+            g_hash_table_insert(stonith_op->meta,
+                                strdup(XML_OP_ATTR_DIGESTS_SECURE),
+                                digests_secure);
         }
 
     } else {
-- 
1.8.3.1

