From fd04f2a9fa8e0bef9bc1dadcc439c59885dff390 Mon Sep 17 00:00:00 2001
From: Ken Gaillot <kgaillot@redhat.com>
Date: Fri, 3 Nov 2017 16:41:29 -0500
Subject: [PATCH] Low: pengine: avoid potential use-of-NULL introduced in rc4

---
 pengine/native.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/pengine/native.c b/pengine/native.c
index 5dc0420..6d1ca24 100644
--- a/pengine/native.c
+++ b/pengine/native.c
@@ -2449,7 +2449,8 @@ LogActions(resource_t * rsc, pe_working_set_t * data_set, gboolean terminal)
                 STOP_SANITY_ASSERT(__LINE__);
             }
 
-            LogAction("Stop", rsc, node, NULL, stop_op, stop_op->reason?stop_op:start, terminal);
+            LogAction("Stop", rsc, node, NULL, stop_op,
+                      (stop_op && stop_op->reason)? stop_op : start, terminal);
         }
 
         free(key);
-- 
1.8.3.1

