From 9b74fb4d667cf187c1c80aeb39ff3b3c12846421 Mon Sep 17 00:00:00 2001
From: Ken Gaillot <kgaillot@redhat.com>
Date: Tue, 18 Apr 2017 14:17:38 -0500
Subject: [PATCH] Low: tools: show Red Hat bugzilla URL when using crm_report

---
 tools/crm_report.in | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/crm_report.in b/tools/crm_report.in
index 26050a7..4715155 100755
--- a/tools/crm_report.in
+++ b/tools/crm_report.in
@@ -222,7 +222,7 @@ EOF
 	log "Collected results are available in $fname"
 	log " "
 	log "Please create a bug entry at"
-	log "    http://bugs.clusterlabs.org/enter_bug.cgi?product=Pacemaker"
+	log "    https://bugzilla.redhat.com/"
 	log "Include a description of your problem and attach this tarball"
 	log " "
 	log "Thank you for taking time to create this report."
-- 
1.8.3.1

