diff -up gsl-1.15/Makefile.in.pom gsl-1.15/Makefile.in
--- gsl-1.15/Makefile.in.pom	2011-05-09 12:37:30.000000000 +0200
+++ gsl-1.15/Makefile.in	2011-05-09 12:48:31.000000000 +0200
@@ -89,9 +89,9 @@ LTLIBRARIES = $(lib_LTLIBRARIES)
 libgsl_la_DEPENDENCIES = $(SUBLIBS) $(am__append_1)
 am_libgsl_la_OBJECTS = version.lo
 libgsl_la_OBJECTS = $(am_libgsl_la_OBJECTS)
-libgsl_la_LINK = $(LIBTOOL) --tag=CC $(AM_LIBTOOLFLAGS) \
-	$(LIBTOOLFLAGS) --mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) \
-	$(libgsl_la_LDFLAGS) $(LDFLAGS) -o $@
+libgsl_la_LINK = $(LIBTOOL) --tag=CC $(AM_LIBTOOLFLAGS)  \
+	$(LIBTOOLFLAGS) --mode=link  $(CCLD)  $(AM_CFLAGS) $(CFLAGS) \
+	$(libgsl_la_LDFLAGS) $(LDFLAGS) -lsatlas -o $@
 PROGRAMS = $(bin_PROGRAMS)
 am_gsl_histogram_OBJECTS = gsl-histogram.$(OBJEXT)
 gsl_histogram_OBJECTS = $(am_gsl_histogram_OBJECTS)
