From a2782063c8daf9000d131e85200bc631a16450b4 Mon Sep 17 00:00:00 2001
From: Rob Clark <robdclark@gmail.com>
Date: Wed, 21 Jun 2017 14:21:15 -0400
Subject: [PATCH 01/23] WIP: fix usb

---
 common/usb_storage.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/common/usb_storage.c b/common/usb_storage.c
index df0b057308..b2a3ab49ec 100644
--- a/common/usb_storage.c
+++ b/common/usb_storage.c
@@ -996,7 +996,7 @@ static int usb_request_sense(struct scsi_cmd *srb, struct us_data *ss)
 
 static int usb_test_unit_ready(struct scsi_cmd *srb, struct us_data *ss)
 {
-	int retries = 10;
+	int retries = 20;
 
 	do {
 		memset(&srb->cmd[0], 0, 12);
@@ -1019,7 +1019,7 @@ static int usb_test_unit_ready(struct scsi_cmd *srb, struct us_data *ss)
 		if ((srb->sense_buf[2] == 0x02) &&
 		    (srb->sense_buf[12] == 0x3a))
 			return -1;
-		mdelay(100);
+		mdelay(250);
 	} while (retries--);
 
 	return -1;
-- 
2.13.3

From 40b06f8d422efc1d9674f081ef22445904c01f4f Mon Sep 17 00:00:00 2001
From: Rob Clark <robdclark@gmail.com>
Date: Mon, 3 Jul 2017 08:34:37 -0400
Subject: [PATCH 02/23] HACK: disable emmc

Hitting some timeout which makes boot take much longer.  And
uefi/boot/rootfs partitions will be on sd-card or usb disk, etc,
so we can just ignore emmc.
---
 arch/arm/dts/dragonboard410c.dts | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/arm/dts/dragonboard410c.dts b/arch/arm/dts/dragonboard410c.dts
index 7746622dda..0d3b7a35f4 100644
--- a/arch/arm/dts/dragonboard410c.dts
+++ b/arch/arm/dts/dragonboard410c.dts
@@ -67,6 +67,7 @@
 			reg = <0x78d9000 0x400>;
 		};
 
+/*
 		sdhci@07824000 {
 			compatible = "qcom,sdhci-msm-v4";
 			reg = <0x7824900 0x11c 0x7824000 0x800>;
@@ -76,6 +77,7 @@
 			clock = <&clkc 0>;
 			clock-frequency = <100000000>;
 		};
+*/
 
 		sdhci@07864000 {
 			compatible = "qcom,sdhci-msm-v4";
-- 
2.13.3
From 7f9d2f7d2ba72cbf62bdabaa98c489b514dc136c Mon Sep 17 00:00:00 2001
From: Peter Robinson <pbrobinson@gmail.com>
Date: Tue, 16 Jan 2018 03:16:35 +0000
Subject: [PATCH 1/2] db410c: config updates

Enable USB keyboards and video output, VFAT writes (useful for EFI boot)
and a common USB eth adapter

Signed-off-by: Peter Robinson <pbrobinson@gmail.com>
---
 configs/dragonboard410c_defconfig | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/configs/dragonboard410c_defconfig b/configs/dragonboard410c_defconfig
index 4389f52b5c..9dcadb938c 100644
--- a/configs/dragonboard410c_defconfig
+++ b/configs/dragonboard410c_defconfig
@@ -41,10 +41,20 @@ CONFIG_USB_HOST_ETHER=y
 CONFIG_USB_ETHER_ASIX=y
 CONFIG_USB_ETHER_ASIX88179=y
 CONFIG_USB_ETHER_MCS7830=y
+CONFIG_USB_ETHER_RTL8152=y
 CONFIG_USB_ETHER_SMSC95XX=y
+CONFIG_DM_ETH=y
 CONFIG_OF_LIBFDT_OVERLAY=y
 CONFIG_OF_CONTROL=y
 CONFIG_ENV_IS_IN_MMC=y
 CONFIG_OF_BOARD_SETUP=y
 CONFIG_PSCI_RESET=y
 CONFIG_OF_SEPARATE=y
+CONFIG_DM_KEYBOARD=y
+CONFIG_USB_KEYBOARD=y
+CONFIG_SYS_USB_EVENT_POLL_VIA_INT_QUEUE=y
+CONFIG_DM_VIDEO=y
+# CONFG_VIDEO_BPP8 is not set
+CONFIG_NO_FB_CLEAR=y
+CONFIG_VIDEO_SIMPLE=y
+CONFIG_FAT_WRITE=y
-- 
2.14.3

From 5d4690815f188a46a2eebad023403819fe4d053a Mon Sep 17 00:00:00 2001
From: Peter Robinson <pbrobinson@gmail.com>
Date: Tue, 16 Jan 2018 03:23:07 +0000
Subject: [PATCH 2/2] db820c: config updates

Enable LEDs, FAT write, and various DT functionality

Signed-off-by: Peter Robinson <pbrobinson@gmail.com>
---
 configs/dragonboard820c_defconfig | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/configs/dragonboard820c_defconfig b/configs/dragonboard820c_defconfig
index 8af54aa3c4..dcec65d212 100644
--- a/configs/dragonboard820c_defconfig
+++ b/configs/dragonboard820c_defconfig
@@ -35,9 +35,15 @@ CONFIG_DM_GPIO=y
 CONFIG_DM_PMIC=y
 CONFIG_PMIC_PM8916=y
 CONFIG_PM8916_GPIO=y
+CONFIG_LED=y
+CONFIG_LED_GPIO=y
 CONFIG_CLK=y
 CONFIG_PSCI_RESET=y
 CONFIG_ENV_IS_IN_EXT4=y
 CONFIG_ENV_EXT4_INTERFACE="mmc"
 CONFIG_ENV_EXT4_DEVICE_AND_PART="0:1"
 CONFIG_ENV_EXT4_FILE="/uboot.env"
+CONFIG_FAT_WRITE=y
+CONFIG_OF_LIBFDT_OVERLAY=y
+CONFIG_OF_CONTROL=y
+CONFIG_OF_SEPARATE=y
-- 
2.14.3

