From dbe4dff86520a1373076563d1bee5ada1d7fd7c5 Mon Sep 17 00:00:00 2001
From: Peter Robinson <pbrobinson@gmail.com>
Date: Thu, 26 Apr 2018 16:56:54 +0100
Subject: [PATCH] tegra186: jetson-tx2: disable onboard emmc

Signed-off-by: Peter Robinson <pbrobinson@gmail.com>
---
 arch/arm/dts/tegra186-p2771-0000.dtsi | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/arm/dts/tegra186-p2771-0000.dtsi b/arch/arm/dts/tegra186-p2771-0000.dtsi
index a1319dc493..e220f086ee 100644
--- a/arch/arm/dts/tegra186-p2771-0000.dtsi
+++ b/arch/arm/dts/tegra186-p2771-0000.dtsi
@@ -53,7 +53,7 @@
 	};
 
 	sdhci@3460000 {
-		status = "okay";
+		status = "disabled";
 		bus-width = <8>;
 		non-removable;
 	};
-- 
2.17.0

