commit b9ede742ef4021380f50f93f33c4013a0fe8d80c
Author: David Smith <dsmith@redhat.com>
Date:   Tue Nov 28 15:35:06 2017 -0500

    rhbz1490862: more kernel f2fs tracepoint header file workarounds

diff --git a/tapsets.cxx b/tapsets.cxx
index f571b88..e1318ec 100644
--- a/tapsets.cxx
+++ b/tapsets.cxx
@@ -11270,6 +11270,9 @@ static vector<string> tracepoint_extra_decls (systemtap_session& s,
       they_live.push_back ("struct f2fs_sb_info;");
       they_live.push_back ("struct extent_info;");
       they_live.push_back ("struct extent_node;");
+      they_live.push_back ("struct super_block;");
+      they_live.push_back ("struct buffer_head;");
+      they_live.push_back ("struct bio;");
     }
 
   if (header.find("radeon") != string::npos)
