From 61223ca7c6ac36c81ff06df502f8c9a02a133708 Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Fri, 13 Sep 2013 10:37:10 -0400
Subject: [PATCH 03/41] Use $(PREFIX) during make install.

Signed-off-by: Peter Jones <pjones@redhat.com>
---
 Makefile | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/Makefile b/Makefile
index 5dbf0f7..94cef27 100644
--- a/Makefile
+++ b/Makefile
@@ -42,14 +42,14 @@ test: all
 	@./test.sh
 
 install: all
-	mkdir -p $(DESTDIR)/sbin
+	mkdir -p $(DESTDIR)$(PREFIX)/sbin
 	mkdir -p $(DESTDIR)/$(mandir)/man8
-	install -m 755 new-kernel-pkg $(DESTDIR)/sbin
+	install -m 755 new-kernel-pkg $(DESTDIR)$(PREFIX)/sbin
 	install -m 644 new-kernel-pkg.8 $(DESTDIR)/$(mandir)/man8
-	install -m 755 installkernel $(DESTDIR)/sbin
+	install -m 755 installkernel $(DESTDIR)$(PREFIX)/sbin
 	install -m 644 installkernel.8 $(DESTDIR)/$(mandir)/man8
 	if [ -f grubby ]; then \
-		install -m 755 grubby $(DESTDIR)/sbin ; \
+		install -m 755 grubby $(DESTDIR)$(PREFIX)/sbin ; \
 		install -m 644 grubby.8 $(DESTDIR)/$(mandir)/man8 ; \
 	fi
 
-- 
2.4.3

