From b8ad9832948f86bf051c15d212fe2c978313c7f8 Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Wed, 12 Feb 2014 10:58:48 -0500
Subject: [PATCH 07/41] Actually /USE/ $DEVTREE in new-kernel-pkg.

Resolves: rhbz#1063534

Signed-off-by: Peter Jones <pjones@redhat.com>
---
 new-kernel-pkg | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/new-kernel-pkg b/new-kernel-pkg
index 7479557..97e1839 100755
--- a/new-kernel-pkg
+++ b/new-kernel-pkg
@@ -230,7 +230,7 @@ install() {
 	[ -n "$verbose" ] && echo "adding $version to $grub2EfiConfig"
 
 	$grubby --grub2 -c $grub2EfiConfig --efi \
-	    --add-kernel=$kernelImage \
+	    --add-kernel=$kernelImage $DEVTREE \
 	    $INITRD --copy-default $makedefault --title "$title" \
 	    ${mbkernel:+--add-multiboot="$mbkernel"} \
 	    ${mbargs:+--mbargs="$mbargs"} \
-- 
2.4.3

