From 0a24262f5c59998ef75d72a1e0a698c33cd8dd28 Mon Sep 17 00:00:00 2001
From: Lubomir Rintel <lkundrak@v3.sk>
Date: Wed, 27 Nov 2013 16:59:47 +0100
Subject: [PATCH 13/41] grub: Remove a redundant test

Signed-off-by: Lubomir Rintel <lkundrak@v3.sk>
---
 test.sh | 2 --
 1 file changed, 2 deletions(-)

diff --git a/test.sh b/test.sh
index 734b755..50a2d25 100755
--- a/test.sh
+++ b/test.sh
@@ -373,8 +373,6 @@ grubTest grub.3 updargs/g3.7 --update-kernel=ALL \
     --remove-args="hdd root ro"
 grubTest grub.7 updargs/g7.2 --boot-filesystem=/    \
     --update-kernel=ALL --args "hde=ide-scsi"
-grubTest grub.7 updargs/g7.2 --boot-filesystem=/    \
-    --update-kernel=ALL --args "hde=ide-scsi"
 grubTest grub.7 updargs/g7.3 --boot-filesystem=/    \
     --update-kernel=DEFAULT --args "hde=ide-scsi"
 grubTest grub.7 updargs/g7.4 --boot-filesystem=/    \
-- 
2.4.3

