From ad075a4b29f5cfd682c65be4f5f4a6e9de9034c4 Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Wed, 21 May 2014 17:37:23 -0400
Subject: [PATCH 18/41] Fix a wrong test case (lacked --boot-filesystem=)

Signed-off-by: Peter Jones <pjones@redhat.com>
---
 test.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/test.sh b/test.sh
index 57e4926..1d95ed3 100755
--- a/test.sh
+++ b/test.sh
@@ -515,7 +515,7 @@ if [ "$testgrub2" == "y" ]; then
         --add-kernel=/boot/vmlinuz-0-rescue-5a94251776a14678911d4ae0949500f5 \
         --initrd /boot/initramfs-0-rescue-5a94251776a14678911d4ae0949500f5.img \
         --copy-default --title "Fedora 21 Rescue" --args=root=/fooooo \
-        --remove-kernel=wtf
+        --remove-kernel=wtf --boot-filesystem=/boot/
 
     testing="GRUB2 add initrd"
     grub2Test grub2.2 add/g2-1.4 --update-kernel=/boot/new-kernel.img \
-- 
2.4.3

