From d4fc18c1a9b9b3e8f8e73afa5c82c938555f5c4d Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Thu, 16 Oct 2014 14:07:01 -0400
Subject: [PATCH 25/41] Get the error checking on getLineByType() return
 /right/.

Signed-off-by: Peter Jones <pjones@redhat.com>
---
 grubby.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/grubby.c b/grubby.c
index febc1ad..82a0f15 100644
--- a/grubby.c
+++ b/grubby.c
@@ -2363,7 +2363,7 @@ void displayEntry(struct singleEntry * entry, const char * prefix, int index) {
     } else {
 	char * title;
 	line = getLineByType(LT_MENUENTRY, entry->lines);
-	if (!line) {
+	if (line) {
 	    title = grub2ExtractTitle(line);
 	    if (title)
 		printf("title=%s\n", title);
-- 
2.4.3

