From 92d813aeec67ee19b7340d3df29347c5b149b8e2 Mon Sep 17 00:00:00 2001
From: "Brian C. Lane" <bcl@redhat.com>
Date: Mon, 13 Apr 2015 13:57:33 -0700
Subject: [PATCH 34/41] Set envFile from --env when bootloader is not specified

---
 grubby.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/grubby.c b/grubby.c
index 5ee20e3..232704a 100644
--- a/grubby.c
+++ b/grubby.c
@@ -4371,9 +4371,11 @@ int main(int argc, const char ** argv) {
     }
 
     if (!cfi) {
-        if (grub2FindConfig(&grub2ConfigType))
+        if (grub2FindConfig(&grub2ConfigType)) {
 	    cfi = &grub2ConfigType;
-	else
+	    if (envPath)
+		cfi->envFile = envPath;
+        } else
       #ifdef __ia64__
 	    cfi = &eliloConfigType;
       #elif __powerpc__
-- 
2.4.3

