From fe41880f07b93914ce8a6132f4e5526c78772935 Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Thu, 2 Jul 2015 12:44:51 -0400
Subject: [PATCH 39/41] Emit better systemd debug settings on debug entries.

Resolves: rhbz#1212128

Signed-off-by: Peter Jones <pjones@redhat.com>
---
 new-kernel-pkg | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/new-kernel-pkg b/new-kernel-pkg
index f6369e4..f547336 100755
--- a/new-kernel-pkg
+++ b/new-kernel-pkg
@@ -121,7 +121,7 @@ mbkernel="$HYPERVISOR"
 mbargs="$HYPERVISOR_ARGS"
 adddracutargs=""
 addplymouthinitrd=""
-DEBUGARG="systemd.debug"
+DEBUGARG="systemd.log_level=debug systemd.log_target=kmsg"
 
 usage() {
     echo "Usage: `basename $0` [-v] [--mkinitrd] [--rminitrd] [--dracut]" >&2
-- 
2.4.3

