From 89e1bba74b80152664bd2ec4d8e45df88051f3bf Mon Sep 17 00:00:00 2001
From: Pablo Greco <pablo@fliagreco.com.ar>
Date: Wed, 21 Mar 2018 12:21:24 +0000
Subject: [PATCH] Update extlinux.conf with sed instead of backporting upstream
 patches

---
 new-kernel-pkg | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/new-kernel-pkg b/new-kernel-pkg
index 847e959..6b8bf8f 100755
--- a/new-kernel-pkg
+++ b/new-kernel-pkg
@@ -308,6 +308,11 @@ install() {
 
 	rungrubby ${ARGS} $makedefault
 	rungrubby --debug ${ARGS} --set-index=1
+	sed -i \
+	    -e '/fdtdir/d' \
+	    -e '/	kernel/{h;s/kernel/fdtdir/g;s/vmlinuz/dtb/g;x}' \
+	    -e '/	append/{p;x}' \
+	    $extlinuxConfig
     else
 	[ -n "$verbose" ] && echo "$extlinuxConfig does not exist, not running grubby for extlinux"
     fi
-- 
1.8.3.1

