From 50d865007906b79e4ba0e5dd717136f17de3713a Mon Sep 17 00:00:00 2001
From: Adam Jackson <ajax@redhat.com>
Date: Fri, 27 Jan 2017 13:38:37 -0500
Subject: [PATCH] Fix ppc64le and arm64

---
 test/dlwrap.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/test/dlwrap.c b/test/dlwrap.c
index 9d1add5..5cda6b9 100644
--- a/test/dlwrap.c
+++ b/test/dlwrap.c
@@ -232,6 +232,7 @@ dlwrap_real_dlsym(void *handle, const char *name)
          * In the meantime, I'll just keep augmenting this
          * hard-coded version list as people report bugs. */
         const char *version[] = {
+            "GLIBC_2.17",
             "GLIBC_2.3",
             "GLIBC_2.2.5",
             "GLIBC_2.2",
-- 
2.9.3

