From 0fcf84cc2af776abfd0ce740e99635c8368c4d91 Mon Sep 17 00:00:00 2001
From: Jarod Wilson <jarod@redhat.com>
Date: Wed, 1 Feb 2017 16:05:39 -0500
Subject: [PATCH rdma-core] redhat/kernel-init: ocrdma is tech-preview too

Signed-off-by: Jarod Wilson <jarod@redhat.com>
---
 redhat/rdma.kernel-init | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/redhat/rdma.kernel-init b/redhat/rdma.kernel-init
index ecfe5b57..d323f830 100644
--- a/redhat/rdma.kernel-init
+++ b/redhat/rdma.kernel-init
@@ -162,10 +162,6 @@ load_hardware_modules()
 	load_modules iw_cxgb4
 	RC+=$?
     fi
-    if is_loaded be2net -a ! is_loaded ocrdma; then
-	load_modules ocrdma
-	RC+=$?
-    fi
     if is_loaded enic -a ! is_loaded usnic_verbs; then
 	load_modules usnic_verbs
 	RC+=$?
@@ -175,6 +171,10 @@ load_hardware_modules()
 	    load_modules i40iw
 	    RC+=$?
         fi
+        if is_loaded be2net -a ! is_loaded ocrdma; then
+	    load_modules ocrdma
+	    RC+=$?
+        fi
     fi
     return $RC
 }
-- 
2.11.0

