--- oldMakefile	2008-07-02 17:48:01.000000000 -0400
+++ openjdk/Makefile	2008-07-02 17:48:09.000000000 -0400
@@ -199,19 +199,19 @@
 
 create_fresh_product_bootdir: FRC
 	$(MAKE) ALT_OUTPUTDIR=$(ABS_BOOTDIR_OUTPUTDIR) \
-		GENERATE_DOCS=false \
+		GENERATE_DOCS=true \
 		BOOT_CYCLE_SETTINGS= \
 		build_product_image
 
 create_fresh_debug_bootdir: FRC
 	$(MAKE) ALT_OUTPUTDIR=$(ABS_BOOTDIR_OUTPUTDIR) \
-		GENERATE_DOCS=false \
+		GENERATE_DOCS=true \
 		BOOT_CYCLE_DEBUG_SETTINGS= \
 		build_debug_image
 
 create_fresh_fastdebug_bootdir: FRC
 	$(MAKE) ALT_OUTPUTDIR=$(ABS_BOOTDIR_OUTPUTDIR) \
-		GENERATE_DOCS=false \
+		GENERATE_DOCS=true \
 		BOOT_CYCLE_DEBUG_SETTINGS= \
 		build_fastdebug_image
 
@@ -262,7 +262,7 @@
 	$(MAKE) \
 		ALT_OUTPUTDIR=$(ABS_OUTPUTDIR)/$(REL_JDK_OUTPUTDIR) \
 	        DEBUG_NAME=$(DEBUG_NAME) \
-		GENERATE_DOCS=false \
+		GENERATE_DOCS=true \
 		$(if $(findstring true,$(BUILD_INSTALL)),BUILD_INSTALL_BUNDLES=true,) \
 		CREATE_DEBUGINFO_BUNDLES=true \
 	        $(BOOT_CYCLE_DEBUG_SETTINGS) \

