--- openjdk/hotspot/build/linux/makefiles/saproc.make_back	2009-12-14 13:35:46.000000000 +0100
+++ openjdk/hotspot/make/linux/makefiles/saproc.make	2009-12-14 13:36:47.000000000 +0100
@@ -95,6 +95,7 @@
 			   $(ALT_SAINCDIR) 										\
 	           $(SASRCFILES)                                        \
 	           $(SA_LFLAGS)                                         \
+	           -g                                                   \
 	           $(SA_DEBUG_CFLAGS)                                   \
 	           $(EXTRA_CFLAGS)                                      \
 	           -o $@                                                \
--- openjdk/hotspot/build/linux/makefiles/jsig.make_back	2009-12-14 13:34:56.000000000 +0100
+++ openjdk/hotspot/make/linux/makefiles/jsig.make	2009-12-14 13:35:31.000000000 +0100
@@ -59,6 +59,7 @@
 $(LIBJSIG): $(JSIGSRCDIR)/jsig.c $(LIBJSIG_MAPFILE)
 	@echo Making signal interposition lib...
 	$(QUIETLY) $(CC) $(SYMFLAG) $(ARCHFLAG) $(SHARED_FLAG) $(PICFLAG) \
+	                     -g							  \
                          $(LFLAGS_JSIG) $(JSIG_DEBUG_CFLAGS) $(EXTRA_CFLAGS) -o $@ $< -ldl
 	$(QUIETLY) [ -f $(LIBJSIG_G) ] || { ln -s $@ $(LIBJSIG_G); }
 ifeq ($(ENABLE_FULL_DEBUG_SYMBOLS),1)
