diff --git a/src/share/lib/security/java.security-linux b/src/share/lib/security/java.security-linux
--- openjdk/jdk/src/share/lib/security/java.security-linux
+++ openjdk/jdk/src/share/lib/security/java.security-linux
@@ -168,6 +168,8 @@
                com.sun.org.glassfish.,\
                jdk.xml.internal.,\
                oracle.jrockit.jfr.,\
+               org.GNOME.Accessibility.,\
+               org.GNOME.Bonobo.,\
                org.jcp.xml.dsig.internal.
 #
 # List of comma-separated packages that start with or equal this string
@@ -211,6 +213,8 @@
                    com.sun.org.glassfish.,\
                    jdk.xml.internal.,\
                    oracle.jrockit.jfr.,\
+                   org.GNOME.Accessibility.,\
+                   org.GNOME.Bonobo.,\
                    org.jcp.xml.dsig.internal.
 #
 # Determines whether this properties file can be appended to
