diff --git a/src/share/lib/security/java.security-linux b/src/share/lib/security/java.security-linux
--- openjdk/jdk/src/share/lib/security/java.security-linux
+++ openjdk/jdk/src/share/lib/security/java.security-linux
@@ -158,6 +158,8 @@
                com.sun.org.apache.xml.internal.utils.,\
                com.sun.org.glassfish.,\
                oracle.jrockit.jfr.,\
+               org.GNOME.Accessibility.,\
+               org.GNOME.Bonobo.,\
                org.jcp.xml.dsig.internal.
 #
 # List of comma-separated packages that start with or equal this string
@@ -200,6 +202,8 @@
                    com.sun.org.apache.xml.internal.utils.,\
                    com.sun.org.glassfish.,\
                    oracle.jrockit.jfr.,\
+                   org.GNOME.Accessibility.,\
+                   org.GNOME.Bonobo.,\
                    org.jcp.xml.dsig.internal.
 #
 # Determines whether this properties file can be appended to
