From 53e3905783593eaa4a06ea3ee416e9dc1da3a60a Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Fri, 24 Apr 2015 09:45:41 -0400
Subject: [PATCH] RHEL 7: Fix list of supported sound cards to match RHEL qemu
 (RHBZ#1176493).

---
 v2v/utils.ml | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/v2v/utils.ml b/v2v/utils.ml
index e53f9fe0a..8f1053ec9 100644
--- a/v2v/utils.ml
+++ b/v2v/utils.ml
@@ -40,13 +40,14 @@ let kvm_arch = function
 (* Does qemu support the given sound card? *)
 let qemu_supports_sound_card = function
   | Types.AC97
-  | Types.ES1370
   | Types.ICH6
   | Types.ICH9
   | Types.PCSpeaker
+    -> true
+  | Types.ES1370
   | Types.SB16
   | Types.USBAudio
-    -> true
+    -> false
 
 (* Find the UEFI firmware. *)
 let find_uefi_firmware guest_arch =
-- 
2.13.4

