From ec0693476e8459771fa74cb820b7f9f89a6ba915 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Mon, 21 Sep 2015 13:12:43 -0400
Subject: [PATCH] RHEL 7: tests: Disable daemon tests that require the 'unix'
 backend.

---
 tests/daemon/Makefile.am | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/tests/daemon/Makefile.am b/tests/daemon/Makefile.am
index 053cad3e1..0d723fee4 100644
--- a/tests/daemon/Makefile.am
+++ b/tests/daemon/Makefile.am
@@ -23,9 +23,7 @@ include $(top_srcdir)/subdir-rules.mk
 
 check_DATA = captive-daemon.pm
 
-TESTS = \
-	test-daemon-start.pl \
-	test-btrfs.pl
+TESTS =
 
 TESTS_ENVIRONMENT = $(top_builddir)/run --test
 
-- 
2.13.4

