diff -up mesa-20140827/src/gallium/drivers/nouveau/codegen/nv50_ir.cpp.fixbuild mesa-20140827/src/gallium/drivers/nouveau/codegen/nv50_ir.cpp
--- mesa-20140827/src/gallium/drivers/nouveau/codegen/nv50_ir.cpp.fixbuild	2014-08-27 15:33:21.858830514 +1000
+++ mesa-20140827/src/gallium/drivers/nouveau/codegen/nv50_ir.cpp	2014-08-27 15:33:33.193830514 +1000
@@ -739,7 +739,7 @@ Instruction::clone(ClonePolicy<Function>
    if (!i)
       i = new_Instruction(pol.context(), op, dType);
 #ifndef NDEBUG // non-conformant assert, so this is required
-   assert(typeid(*i) == typeid(*this));
+   //assert(typeid(*i) == typeid(*this));
 #endif
 
    pol.set<Instruction>(this, i);
