From 9ec76750dc457e0a445fd43bbc172ab3c3de1a23 Mon Sep 17 00:00:00 2001
From: Tomas Jelinek <tojeline@redhat.com>
Date: Thu, 1 Feb 2018 11:10:27 +0100
Subject: [PATCH 1/2] fix 'pcs cluster auth'

---
 pcsd/cfgsync.rb | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pcsd/cfgsync.rb b/pcsd/cfgsync.rb
index 14b55736..36ae1abb 100644
--- a/pcsd/cfgsync.rb
+++ b/pcsd/cfgsync.rb
@@ -795,7 +795,7 @@ module Cfgsync
     # we run in a cluster so we need to sync the config
     publisher = ConfigPublisher.new(
       PCSAuth.getSuperuserAuth(), [config_new], nodes, cluster_name,
-      new_tokens
+      new_tokens, new_ports
     )
     old_configs, node_responses = publisher.publish()
     if not old_configs.include?(config_new.class.name)
-- 
2.13.6

