From 4c2dc964999f57395f377d44594324e5ae3df1b0 Mon Sep 17 00:00:00 2001
From: Tomas Jelinek <tojeline@redhat.com>
Date: Thu, 1 Feb 2018 13:13:23 +0100
Subject: [PATCH 2/2] fix 'pcs cluster auth'

---
 pcsd/pcs.rb | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pcsd/pcs.rb b/pcsd/pcs.rb
index 68d2e7ed..43f58f22 100644
--- a/pcsd/pcs.rb
+++ b/pcsd/pcs.rb
@@ -1247,7 +1247,7 @@ def pcs_auth(auth_user, nodes, username, password, force=false, local=true)
   auth_responses.each { |node, response|
     if 'ok' == response['status']
       new_tokens[node] = response['token']
-      ports[node] = nodes[node]
+      ports[node] = nodes[node] || PCSD_DEFAULT_PORT
     end
   }
   if not new_tokens.empty?
-- 
2.13.6

