From 772b48493eb19f2d9e25579740edd6216c9f8ed0 Mon Sep 17 00:00:00 2001
From: Tomas Jelinek <tojeline@redhat.com>
Date: Thu, 4 Jan 2018 13:22:03 +0100
Subject: [PATCH] fix a crash in 'pcs booth sync'

---
 pcs/lib/communication/booth.py | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/pcs/lib/communication/booth.py b/pcs/lib/communication/booth.py
index 240738a..a3aa918 100644
--- a/pcs/lib/communication/booth.py
+++ b/pcs/lib/communication/booth.py
@@ -10,6 +10,7 @@ import os
 
 from pcs.common.node_communicator import RequestData
 from pcs.lib import reports
+from pcs.lib.booth import reports as reports_booth
 from pcs.lib.communication.tools import (
     AllAtOnceStrategyMixin,
     AllSameDataMixin,
@@ -51,12 +52,12 @@ class BoothSendConfig(
         )
 
     def _get_success_report(self, node_label):
-        return reports.booth_config_accepted_by_node(
+        return reports_booth.booth_config_accepted_by_node(
             node_label, [self._booth_name]
         )
 
     def before(self):
-        self._report(reports.booth_config_distribution_started())
+        self._report(reports_booth.booth_config_distribution_started())
 
 
 class ProcessJsonDataMixin(object):
-- 
1.8.3.1

