From 51485a618da52af2c6bb114e70a52a51b68e6a09 Mon Sep 17 00:00:00 2001
From: Tomas Jelinek <tojeline@redhat.com>
Date: Tue, 21 Nov 2017 17:05:35 +0100
Subject: [PATCH] fix --skip-offline in 'pcs quorum device remove'

---
 pcs/lib/commands/quorum.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pcs/lib/commands/quorum.py b/pcs/lib/commands/quorum.py
index 3e9db0e..6b869ed 100644
--- a/pcs/lib/commands/quorum.py
+++ b/pcs/lib/commands/quorum.py
@@ -325,7 +325,7 @@ def _remove_device_model_net(lib_env, cluster_nodes, skip_offline_nodes):
     reporter.process(
         reports.qdevice_certificate_removal_started()
     )
-    com_cmd = qdevice_net_com.ClientDestroy(reporter)
+    com_cmd = qdevice_net_com.ClientDestroy(reporter, skip_offline_nodes)
     com_cmd.set_targets(
         lib_env.get_node_target_factory().get_target_list(cluster_nodes)
     )
-- 
1.8.3.1

