--- mariadb-5.5.40/scripts/mysql_config.sh.basedir	2014-10-08 15:19:53.000000000 +0200
+++ mariadb-5.5.40/scripts/mysql_config.sh	2014-11-06 13:36:54.025734370 +0100
@@ -77,7 +77,7 @@
 me=`get_full_path $0`
 
 # Script might have been renamed but assume mysql_<something>config<something>
-basedir=`echo $me | sed -e 's;/bin/mysql_.*config.*;;'`
+basedir='@prefix@'
 
 ldata='@localstatedir@'
 execdir='@libexecdir@'
