--- valgrind-3.10.0.BETA1/configure.orig	2014-09-02 13:49:58.125269536 +0200
+++ valgrind-3.10.0.BETA1/configure	2014-09-02 13:54:08.188701144 +0200
@@ -8978,9 +8978,8 @@
 fi
 rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
 if test x$ac_have_as_ppc_fpPO = xyes ; then
-
-$as_echo "#define HAVE_AS_PPC_FPPO 1" >>confdefs.h
-
+  { $as_echo "$as_me:${as_lineno-$LINENO}: HAVE_AS_PPC_FPPO found, but ignoring." >&5
+$as_echo "$as_me: HAVE_AS_PPC_FPPO found, but ignoring." >&6;}
 fi
 
 
--- valgrind-3.11.0/config.h.in.orig	2016-04-15 23:51:26.889440687 +0200
+++ valgrind-3.11.0/config.h.in	2016-04-15 23:51:43.430126014 +0200
@@ -55,9 +55,6 @@
 /* Define to 1 if as supports fxsave64/fxrstor64. */
 #undef HAVE_AS_AMD64_FXSAVE64
 
-/* Define to 1 if as supports floating point phased out category. */
-#undef HAVE_AS_PPC_FPPO
-
 /* Define to 1 if as supports mtocrf/mfocrf. */
 #undef HAVE_AS_PPC_MFTOCRF
 
--- valgrind-3.10.0.BETA1/configure.ac.orig	2014-09-02 14:01:04.443754937 +0200
+++ valgrind-3.10.0.BETA1/configure.ac	2014-09-02 14:01:16.483727639 +0200
@@ -2022,7 +2022,7 @@
 AC_MSG_RESULT([no])
 ])
 if test x$ac_have_as_ppc_fpPO = xyes ; then
-  AC_DEFINE(HAVE_AS_PPC_FPPO, 1, [Define to 1 if as supports floating point phased out category.])
+  AC_MSG_NOTICE([HAVE_AS_PPC_FPPO found, but ignoring.])
 fi
 
 
