diff --git a/coregrind/m_syswrap/syswrap-arm64-linux.c b/coregrind/m_syswrap/syswrap-arm64-linux.c
index f8b5b7c..60482d7 100644
--- a/coregrind/m_syswrap/syswrap-arm64-linux.c
+++ b/coregrind/m_syswrap/syswrap-arm64-linux.c
@@ -666,10 +666,11 @@ static SyscallTableEntry syscall_main_table[] = {
    LINX_(__NR_unlinkat,          sys_unlinkat),          // 35
    LINX_(__NR_symlinkat,         sys_symlinkat),         // 36
    LINX_(__NR_linkat,            sys_linkat),            // 37
-   LINX_(__NR_renameat,		 sys_renameat),          // 38
+   LINX_(__NR_renameat,          sys_renameat),          // 38
    LINX_(__NR_umount2,           sys_umount),            // 39
    LINX_(__NR_mount,             sys_mount),             // 40
-
+   LINX_(__NR_pivot_root,        sys_pivot_root),        // 41
+   //   (__NR_nfsservctl,        sys_ni_syscall),        // 42
    GENXY(__NR_statfs,            sys_statfs),            // 43
    GENXY(__NR_fstatfs,           sys_fstatfs),           // 44
    GENX_(__NR_truncate,          sys_truncate),          // 45
@@ -711,7 +712,7 @@ static SyscallTableEntry syscall_main_table[] = {
    GENX_(__NR_sync,              sys_sync),              // 81
    GENX_(__NR_fsync,             sys_fsync),             // 82
    GENX_(__NR_fdatasync,         sys_fdatasync),         // 83
-
+   LINX_(__NR_sync_file_range,   sys_sync_file_range),   // 84
    LINXY(__NR_timerfd_create,    sys_timerfd_create),    // 85
    LINXY(__NR_timerfd_settime,   sys_timerfd_settime),   // 86
    LINXY(__NR_timerfd_gettime,   sys_timerfd_gettime),   // 87
@@ -724,27 +725,27 @@ static SyscallTableEntry syscall_main_table[] = {
    LINX_(__NR_exit_group,        sys_exit_group),        // 94
    LINXY(__NR_waitid,            sys_waitid),            // 95
    LINX_(__NR_set_tid_address,   sys_set_tid_address),   // 96
-
+   LINX_(__NR_unshare,           sys_unshare),           // 97
    LINXY(__NR_futex,             sys_futex),             // 98
    LINX_(__NR_set_robust_list,   sys_set_robust_list),   // 99
-
+   LINXY(__NR_get_robust_list,   sys_get_robust_list),   // 100
    GENXY(__NR_nanosleep,         sys_nanosleep),         // 101
    GENXY(__NR_getitimer,         sys_getitimer),         // 102
    GENXY(__NR_setitimer,         sys_setitimer),         // 103
    GENX_(__NR_kexec_load,        sys_ni_syscall),        // 104
    LINX_(__NR_init_module,       sys_init_module),       // 105
-
+   LINX_(__NR_delete_module,     sys_delete_module),     // 106
    LINXY(__NR_timer_create,      sys_timer_create),      // 107
-   LINXY(__NR_timer_settime,     sys_timer_settime),     // 108
-   LINXY(__NR_timer_gettime,     sys_timer_gettime),     // 109
-   LINX_(__NR_timer_getoverrun,  sys_timer_getoverrun),  // 110
+   LINXY(__NR_timer_gettime,     sys_timer_gettime),     // 108
+   LINX_(__NR_timer_getoverrun,  sys_timer_getoverrun),  // 109
+   LINXY(__NR_timer_settime,     sys_timer_settime),     // 110
    LINX_(__NR_timer_delete,      sys_timer_delete),      // 111
    LINX_(__NR_clock_settime,     sys_clock_settime),     // 112
    LINXY(__NR_clock_gettime,     sys_clock_gettime),     // 113
    LINXY(__NR_clock_getres,      sys_clock_getres),      // 114
    LINXY(__NR_clock_nanosleep,   sys_clock_nanosleep),   // 115
    LINXY(__NR_syslog,            sys_syslog),            // 116
-
+   //   (__NR_ptrace,            sys_ptrace),            // 117
    LINXY(__NR_sched_setparam,    sys_sched_setparam),    // 118
    LINX_(__NR_sched_setscheduler,sys_sched_setscheduler),// 119
    LINX_(__NR_sched_getscheduler,sys_sched_getscheduler),// 120
@@ -754,9 +755,10 @@ static SyscallTableEntry syscall_main_table[] = {
    LINX_(__NR_sched_yield,       sys_sched_yield),       // 124
    LINX_(__NR_sched_get_priority_max, sys_sched_get_priority_max),// 125
    LINX_(__NR_sched_get_priority_min, sys_sched_get_priority_min),// 126
-
+   LINXY(__NR_sched_rr_get_interval,   sys_sched_rr_get_interval),// 127
+   //   (__NR_restart_syscall,   sys_ni_syscall),        // 128
    GENX_(__NR_kill,              sys_kill),              // 129
-
+   LINXY(__NR_tkill,             sys_tkill),             // 130
    LINX_(__NR_tgkill,            sys_tgkill),            // 131
    GENXY(__NR_sigaltstack,       sys_sigaltstack),       // 132
    LINX_(__NR_rt_sigsuspend,     sys_rt_sigsuspend),     // 133
@@ -768,7 +770,7 @@ static SyscallTableEntry syscall_main_table[] = {
    PLAX_(__NR_rt_sigreturn,      sys_rt_sigreturn),      // 139
    GENX_(__NR_setpriority,       sys_setpriority),       // 140
    GENX_(__NR_getpriority,       sys_getpriority),       // 141
-
+   //   (__NR_reboot,            sys_ni_syscall),        // 142
    GENX_(__NR_setregid,          sys_setregid),          // 143
    GENX_(__NR_setgid,            sys_setgid),            // 144
    GENX_(__NR_setreuid,          sys_setreuid),          // 145
@@ -787,13 +789,14 @@ static SyscallTableEntry syscall_main_table[] = {
    GENXY(__NR_getgroups,         sys_getgroups),         // 158
    GENX_(__NR_setgroups,         sys_setgroups),         // 159
    GENXY(__NR_uname,             sys_newuname),          // 160
-
+   GENX_(__NR_sethostname,       sys_sethostname),       // 161
+   //   (__NR_setdomainname,     sys_ni_syscall),        // 162
    GENXY(__NR_getrlimit,         sys_old_getrlimit),     // 163
    GENX_(__NR_setrlimit,         sys_setrlimit),         // 164
    GENXY(__NR_getrusage,         sys_getrusage),         // 165
    GENX_(__NR_umask,             sys_umask),             // 166
    LINXY(__NR_prctl,             sys_prctl),             // 167 
-
+   LINXY(__NR_getcpu,            sys_getcpu),            // 168
    GENXY(__NR_gettimeofday,      sys_gettimeofday),      // 169
    GENX_(__NR_settimeofday,      sys_settimeofday),      // 170
    LINXY(__NR_adjtimex,          sys_adjtimex),          // 171
@@ -843,13 +846,14 @@ static SyscallTableEntry syscall_main_table[] = {
    GENXY(__NR_munmap,            sys_munmap),            // 215
    GENX_(__NR_mremap,            sys_mremap),            // 216
    LINX_(__NR_add_key,           sys_add_key),           // 217
-
+   LINX_(__NR_request_key,       sys_request_key),       // 218
    LINXY(__NR_keyctl,            sys_keyctl),            // 219
    LINX_(__NR_clone,             sys_clone),             // 220
    GENX_(__NR_execve,            sys_execve),            // 221
    PLAX_(__NR_mmap,              sys_mmap),              // 222
    PLAX_(__NR_fadvise64,         sys_fadvise64),         // 223
-
+   //   (__NR_swapon,            sys_swapon),            // 224
+   //   (__NR_swapoff,           sys_swapoff),           // 225
    GENXY(__NR_mprotect,          sys_mprotect),          // 226
    GENX_(__NR_msync,             sys_msync),             // 227
    GENX_(__NR_mlock,             sys_mlock),             // 228
@@ -858,301 +862,47 @@ static SyscallTableEntry syscall_main_table[] = {
    LINX_(__NR_munlockall,        sys_munlockall),        // 231
    GENXY(__NR_mincore,           sys_mincore),           // 232
    GENX_(__NR_madvise,           sys_madvise),           // 233
-
+   //   (__NR_remap_file_pages,  sys_ni_syscall)         // 234
    LINX_(__NR_mbind,             sys_mbind),             // 235
    LINXY(__NR_get_mempolicy,     sys_get_mempolicy),     // 236
    LINX_(__NR_set_mempolicy,     sys_set_mempolicy),     // 237
-
+   //   (__NR_migrate_pages,     sys_ni_syscall),        // 238
+   LINXY(__NR_move_pages,        sys_move_pages),        // 239
+   LINXY(__NR_rt_tgsigqueueinfo, sys_rt_tgsigqueueinfo), // 240
    LINXY(__NR_perf_event_open,   sys_perf_event_open),   // 241
    LINXY(__NR_accept4,           sys_accept4),           // 242
    LINXY(__NR_recvmmsg,          sys_recvmmsg),          // 243
-
    GENXY(__NR_wait4,             sys_wait4),             // 260
    LINXY(__NR_prlimit64,         sys_prlimit64),         // 261
-
+   LINXY(__NR_fanotify_init,     sys_fanotify_init),     // 262
+   LINX_(__NR_fanotify_mark,     sys_fanotify_mark),     // 263
    LINXY(__NR_name_to_handle_at, sys_name_to_handle_at), // 264
    LINXY(__NR_open_by_handle_at, sys_open_by_handle_at), // 265
-
+   LINXY(__NR_clock_adjtime,     sys_clock_adjtime),     // 266
    LINX_(__NR_syncfs,            sys_syncfs),            // 267
-
+   //   (__NR_setns,             sys_ni_syscall),        // 268
    LINXY(__NR_sendmmsg,          sys_sendmmsg),          // 269
    LINXY(__NR_process_vm_readv,  sys_process_vm_readv),  // 270
    LINX_(__NR_process_vm_writev, sys_process_vm_writev), // 271
-
+   LINX_(__NR_kcmp,              sys_kcmp),              // 272
+   //   (__NR_finit_module,      sys_ni_syscall),        // 273
+   //   (__NR_sched_setattr,     sys_ni_syscall),        // 274
+   //   (__NR_sched_getattr,     sys_ni_syscall),        // 275
    LINX_(__NR_renameat2,         sys_renameat2),         // 276
-
+   //   (__NR_seccomp,           sys_ni_syscall),        // 277
    LINXY(__NR_getrandom,         sys_getrandom),         // 278
    LINXY(__NR_memfd_create,      sys_memfd_create),      // 279
-
-// The numbers below are bogus.  (See comment further down.)
-// When pulling entries above this line, change the numbers
-// to be correct.
-
-//ZZ //zz    //   (restart_syscall)                             // 0
-//ZZ    GENX_(__NR_fork,              sys_fork),           // 2
-//ZZ 
-//ZZ    GENXY(__NR_open,              sys_open),           // 5
-//ZZ //   GENXY(__NR_waitpid,           sys_waitpid),        // 7
-//ZZ    GENXY(__NR_creat,             sys_creat),          // 8
-//ZZ    GENX_(__NR_link,              sys_link),           // 9
-//ZZ 
-//ZZ    GENX_(__NR_unlink,            sys_unlink),         // 10
-//ZZ    GENXY(__NR_time,              sys_time),           // 13
-//ZZ    GENX_(__NR_mknod,             sys_mknod),          // 14
-//ZZ 
-//ZZ    GENX_(__NR_chmod,             sys_chmod),          // 15
-//ZZ //zz    LINX_(__NR_lchown,            sys_lchown16),       // 16
-//ZZ //   GENX_(__NR_break,             sys_ni_syscall),     // 17
-//ZZ //zz    //   (__NR_oldstat,           sys_stat),           // 18 (obsolete)
-//ZZ    LINX_(__NR_lseek,             sys_lseek),          // 19
-//ZZ 
-//ZZ    GENX_(__NR_getpid,            sys_getpid),         // 20
-//ZZ    LINX_(__NR_umount,            sys_oldumount),      // 22
-//ZZ    LINX_(__NR_setuid,            sys_setuid16),       // 23 ## P
-//ZZ    LINX_(__NR_getuid,            sys_getuid16),       // 24 ## P
-//ZZ //zz 
-//ZZ //zz    //   (__NR_stime,             sys_stime),          // 25 * (SVr4,SVID,X/OPEN)
-//ZZ    PLAXY(__NR_ptrace,            sys_ptrace),         // 26
-//ZZ    GENX_(__NR_alarm,             sys_alarm),          // 27
-//ZZ //zz    //   (__NR_oldfstat,          sys_fstat),          // 28 * L -- obsolete
-//ZZ    GENX_(__NR_pause,             sys_pause),          // 29
-//ZZ 
-//ZZ    LINX_(__NR_utime,             sys_utime),          // 30
-//ZZ //   GENX_(__NR_stty,              sys_ni_syscall),     // 31
-//ZZ //   GENX_(__NR_gtty,              sys_ni_syscall),     // 32
-//ZZ    GENX_(__NR_access,            sys_access),         // 33
-//ZZ    GENX_(__NR_nice,              sys_nice),           // 34
-//ZZ 
-//ZZ //   GENX_(__NR_ftime,             sys_ni_syscall),     // 35
-//ZZ    GENX_(__NR_rename,            sys_rename),         // 38
-//ZZ    GENX_(__NR_mkdir,             sys_mkdir),          // 39
-//ZZ 
-//ZZ    GENX_(__NR_rmdir,             sys_rmdir),          // 40
-//ZZ    LINXY(__NR_pipe,              sys_pipe),           // 42
-//ZZ //   GENX_(__NR_prof,              sys_ni_syscall),     // 44
-
-//ZZ    LINX_(__NR_getgid,            sys_getgid16),       // 47
-//ZZ //zz    //   (__NR_signal,            sys_signal),         // 48 */* (ANSI C)
-//ZZ    LINX_(__NR_geteuid,           sys_geteuid16),      // 49
-//ZZ 
-//ZZ    LINX_(__NR_getegid,           sys_getegid16),      // 50
-//ZZ //   GENX_(__NR_lock,              sys_ni_syscall),     // 53
-//ZZ 
-//ZZ    LINXY(__NR_fcntl,             sys_fcntl),          // 55
-//ZZ //   GENX_(__NR_mpx,               sys_ni_syscall),     // 56
-//ZZ //   GENX_(__NR_ulimit,            sys_ni_syscall),     // 58
-//ZZ //zz    //   (__NR_oldolduname,       sys_olduname),       // 59 Linux -- obsolete
-//ZZ //zz 
-//ZZ //zz    //   (__NR_ustat,             sys_ustat)           // 62 SVr4 -- deprecated
-//ZZ    GENXY(__NR_dup2,              sys_dup2),           // 63
-//ZZ    GENX_(__NR_getppid,           sys_getppid),        // 64
-//ZZ 
-//ZZ    GENX_(__NR_getpgrp,           sys_getpgrp),        // 65
-//ZZ    LINXY(__NR_sigaction,         sys_sigaction),      // 67
-//ZZ //zz    //   (__NR_sgetmask,          sys_sgetmask),       // 68 */* (ANSI C)
-//ZZ //zz    //   (__NR_ssetmask,          sys_ssetmask),       // 69 */* (ANSI C)
-//ZZ //zz 
-//ZZ    PLAX_(__NR_sigsuspend,        sys_sigsuspend),     // 72
-//ZZ    LINXY(__NR_sigpending,        sys_sigpending),     // 73
-//ZZ //zz    //   (__NR_sethostname,       sys_sethostname),    // 74 */*
-//ZZ //zz 
-//ZZ    GENXY(__NR_getrlimit,         sys_old_getrlimit),  // 76
-//ZZ 
-//ZZ    LINXY(__NR_getgroups,         sys_getgroups16),    // 80
-//ZZ    LINX_(__NR_setgroups,         sys_setgroups16),    // 81
-//ZZ //   PLAX_(__NR_select,            old_select),         // 82
-//ZZ    GENX_(__NR_symlink,           sys_symlink),        // 83
-//ZZ //zz    //   (__NR_oldlstat,          sys_lstat),          // 84 -- obsolete
-//ZZ //zz 
-//ZZ    GENX_(__NR_readlink,          sys_readlink),       // 85
-//ZZ //zz    //   (__NR_uselib,            sys_uselib),         // 86 */Linux
-//ZZ //zz    //   (__NR_swapon,            sys_swapon),         // 87 */Linux
-//ZZ //zz    //   (__NR_reboot,            sys_reboot),         // 88 */Linux
-//ZZ //zz    //   (__NR_readdir,           old_readdir),        // 89 -- superseded
-//ZZ //zz 
-//ZZ //   _____(__NR_mmap,              old_mmap),           // 90
-//ZZ    GENXY(__NR_munmap,            sys_munmap),         // 91
-//ZZ    GENX_(__NR_truncate,          sys_truncate),       // 92
-//ZZ    GENX_(__NR_ftruncate,         sys_ftruncate),      // 93
-//ZZ 
-//ZZ    LINX_(__NR_fchown,            sys_fchown16),       // 95
-//ZZ //   GENX_(__NR_profil,            sys_ni_syscall),     // 98
-//ZZ    GENXY(__NR_statfs,            sys_statfs),         // 99
-//ZZ 
-//ZZ    GENXY(__NR_fstatfs,           sys_fstatfs),        // 100
-//ZZ //   LINX_(__NR_ioperm,            sys_ioperm),         // 101
-//ZZ    LINXY(__NR_socketcall,        sys_socketcall),     // 102
-//ZZ 
-//ZZ    GENXY(__NR_stat,              sys_newstat),        // 106
-//ZZ    GENXY(__NR_lstat,             sys_newlstat),       // 107
-//ZZ    GENXY(__NR_fstat,             sys_newfstat),       // 108
-//ZZ //zz    //   (__NR_olduname,          sys_uname),          // 109 -- obsolete
-//ZZ //zz 
-//ZZ //   GENX_(__NR_iopl,              sys_iopl),           // 110
-//ZZ //   GENX_(__NR_idle,              sys_ni_syscall),     // 112
-//ZZ // PLAXY(__NR_vm86old,           sys_vm86old),        // 113 __NR_syscall... weird
-//ZZ //zz 
-//ZZ //zz    //   (__NR_swapoff,           sys_swapoff),        // 115 */Linux 
-//ZZ //   _____(__NR_ipc,               sys_ipc),            // 117
-//ZZ    GENX_(__NR_fsync,             sys_fsync),          // 118
-//ZZ    PLAX_(__NR_sigreturn,         sys_sigreturn),      // 119 ?/Linux
-//ZZ 
-//ZZ //zz    //   (__NR_setdomainname,     sys_setdomainname),  // 121 */*(?)
-//ZZ //   PLAX_(__NR_modify_ldt,        sys_modify_ldt),     // 123
-//ZZ //zz 
-//ZZ    LINXY(__NR_sigprocmask,       sys_sigprocmask),    // 126
-//ZZ //zz    // Nb: create_module() was removed 2.4-->2.6
-//ZZ //   GENX_(__NR_create_module,     sys_ni_syscall),     // 127
-//ZZ    LINX_(__NR_delete_module,     sys_delete_module),  // 129
-//ZZ //zz 
-//ZZ //zz    // Nb: get_kernel_syms() was removed 2.4-->2.6
-//ZZ //   GENX_(__NR_get_kernel_syms,   sys_ni_syscall),     // 130
-//ZZ    GENX_(__NR_getpgid,           sys_getpgid),        // 132
-//ZZ //zz    //   (__NR_bdflush,           sys_bdflush),        // 134 */Linux
-//ZZ //zz 
-//ZZ //zz    //   (__NR_sysfs,             sys_sysfs),          // 135 SVr4
-//ZZ //   GENX_(__NR_afs_syscall,       sys_ni_syscall),     // 137
-//ZZ  
-//ZZ    LINXY(__NR__llseek,           sys_llseek),         // 140
-//ZZ    GENXY(__NR_getdents,          sys_getdents),       // 141
-//ZZ    GENX_(__NR__newselect,        sys_select),         // 142
-//ZZ 
-//ZZ    LINXY(__NR__sysctl,           sys_sysctl),         // 149
-//ZZ 
-//ZZ //zz    //LINX?(__NR_sched_rr_get_interval,  sys_sched_rr_get_interval), // 161 */*
-//ZZ    LINX_(__NR_setresuid,         sys_setresuid16),    // 164
-//ZZ 
-//ZZ    LINXY(__NR_getresuid,         sys_getresuid16),    // 165
-//ZZ //   PLAXY(__NR_vm86,              sys_vm86),           // 166 x86/Linux-only
-//ZZ //   GENX_(__NR_query_module,      sys_ni_syscall),     // 167
-//ZZ    GENXY(__NR_poll,              sys_poll),           // 168
-//ZZ //zz    //   (__NR_nfsservctl,        sys_nfsservctl),     // 169 */Linux
-//ZZ //zz 
-//ZZ    LINX_(__NR_setresgid,         sys_setresgid16),    // 170
-//ZZ    LINXY(__NR_getresgid,         sys_getresgid16),    // 171
-//ZZ    LINXY(__NR_prctl,             sys_prctl),          // 172
-//ZZ    LINXY(__NR_rt_sigaction,      sys_rt_sigaction),   // 174
-//ZZ 
-//ZZ    LINXY(__NR_rt_sigtimedwait,   sys_rt_sigtimedwait),// 177
-//ZZ 
-//ZZ    LINX_(__NR_chown,             sys_chown16),        // 182
-//ZZ 
-//ZZ    LINXY(__NR_sendfile,          sys_sendfile),       // 187
-//ZZ //   GENXY(__NR_getpmsg,           sys_getpmsg),        // 188
-//ZZ //   GENX_(__NR_putpmsg,           sys_putpmsg),        // 189
-//ZZ 
-//ZZ    // Nb: we treat vfork as fork
-//ZZ    GENX_(__NR_vfork,             sys_fork),           // 190
-//ZZ    GENXY(__NR_ugetrlimit,        sys_getrlimit),      // 191
-//ZZ    GENX_(__NR_truncate64,        sys_truncate64),     // 193
-//ZZ    GENX_(__NR_ftruncate64,       sys_ftruncate64),    // 194
-//ZZ    
-//ZZ    PLAXY(__NR_stat64,            sys_stat64),         // 195
-//ZZ    PLAXY(__NR_lstat64,           sys_lstat64),        // 196
-//ZZ    PLAXY(__NR_fstat64,           sys_fstat64),        // 197
-//ZZ    GENX_(__NR_lchown32,          sys_lchown),         // 198
-//ZZ    GENX_(__NR_getuid32,          sys_getuid),         // 199
-//ZZ 
-//ZZ    GENX_(__NR_getgid32,          sys_getgid),         // 200
-//ZZ    GENX_(__NR_geteuid32,         sys_geteuid),        // 201
-//ZZ    GENX_(__NR_getegid32,         sys_getegid),        // 202
-//ZZ    GENX_(__NR_setreuid32,        sys_setreuid),       // 203
-//ZZ    GENX_(__NR_setregid32,        sys_setregid),       // 204
-//ZZ 
-//ZZ    LINX_(__NR_setresuid32,       sys_setresuid),      // 208
-//ZZ    LINXY(__NR_getresuid32,       sys_getresuid),      // 209
-//ZZ 
-//ZZ    LINX_(__NR_setresgid32,       sys_setresgid),      // 210
-//ZZ    LINXY(__NR_getresgid32,       sys_getresgid),      // 211
-//ZZ    GENX_(__NR_chown32,           sys_chown),          // 212
-//ZZ    GENX_(__NR_setuid32,          sys_setuid),         // 213
-//ZZ    GENX_(__NR_setgid32,          sys_setgid),         // 214
-//ZZ 
-//ZZ    LINX_(__NR_setfsuid32,        sys_setfsuid),       // 215
-//ZZ    LINX_(__NR_setfsgid32,        sys_setfsgid),       // 216
-//ZZ //zz    //   (__NR_pivot_root,        sys_pivot_root),     // 217 */Linux
-//ZZ 
-//ZZ    LINXY(__NR_fcntl64,           sys_fcntl64),        // 221
-//ZZ //   GENX_(222,                    sys_ni_syscall),     // 222
-//ZZ //   PLAXY(223,                    sys_syscall223),     // 223 // sys_bproc?
-//ZZ 
-//ZZ    LINXY(__NR_tkill,             sys_tkill),          // 238 */Linux
-//ZZ 
-//ZZ    LINXY(__NR_futex,             sys_futex),             // 240
-//ZZ    LINXY(__NR_sched_getaffinity, sys_sched_getaffinity), // 242
-//ZZ //   PLAX_(__NR_set_thread_area,   sys_set_thread_area),   // 243
-//ZZ //   PLAX_(__NR_get_thread_area,   sys_get_thread_area),   // 244
-//ZZ 
-//ZZ //   LINX_(__NR_fadvise64,         sys_fadvise64),      // 250 */(Linux?)
-//ZZ    GENX_(251,                    sys_ni_syscall),     // 251
-//ZZ    LINXY(__NR_epoll_create,      sys_epoll_create),   // 254
-//ZZ 
-//ZZ    LINX_(__NR_epoll_ctl,         sys_epoll_ctl),         // 255
-//ZZ    LINXY(__NR_epoll_wait,        sys_epoll_wait),        // 256
-//ZZ //zz    //   (__NR_remap_file_pages,  sys_remap_file_pages),  // 257 */Linux
-//ZZ    LINX_(__NR_set_tid_address,   sys_set_tid_address),   // 258
-//ZZ 
-//ZZ    LINXY(__NR_clock_getres,      sys_clock_getres),   // (timer_create+7)
-//ZZ    GENXY(__NR_statfs64,          sys_statfs64),       // 268
-//ZZ    GENXY(__NR_fstatfs64,         sys_fstatfs64),      // 269
-//ZZ 
-//ZZ    GENX_(__NR_utimes,            sys_utimes),         // 271
-//ZZ //   LINX_(__NR_fadvise64_64,      sys_fadvise64_64),   // 272 */(Linux?)
-//ZZ    GENX_(__NR_vserver,           sys_ni_syscall),     // 273
-//ZZ    LINX_(__NR_mbind,             sys_mbind),          // 274 ?/?
-//ZZ 
-//ZZ    LINXY(__NR_get_mempolicy,     sys_get_mempolicy),  // 275 ?/?
-//ZZ    LINX_(__NR_set_mempolicy,     sys_set_mempolicy),  // 276 ?/?
-//ZZ 
-//ZZ    LINX_(__NR_send,              sys_send),
-//ZZ    LINXY(__NR_recv,              sys_recv),
-//ZZ    LINXY(__NR_recvfrom,          sys_recvfrom),       // 292
-//ZZ    LINX_(__NR_semget,            sys_semget),         // 299
-//ZZ    LINXY(__NR_semctl,            sys_semctl),         // 300
-//ZZ 
-//ZZ    LINX_(__NR_request_key,       sys_request_key),    // 287
-//ZZ    LINX_(__NR_inotify_init,    sys_inotify_init),   // 291
-//ZZ //   LINX_(__NR_migrate_pages,    sys_migrate_pages),    // 294
-//ZZ 
-//ZZ    LINX_(__NR_futimesat,    sys_futimesat),        // 326 on arm
-//ZZ 
-//ZZ    PLAXY(__NR_fstatat64,    sys_fstatat64),        // 300
-//ZZ    LINX_(__NR_renameat,       sys_renameat),         // 302
-//ZZ    LINX_(__NR_symlinkat,    sys_symlinkat),        // 304
-//ZZ 
-//ZZ    LINX_(__NR_shmget,            sys_shmget),         //307 
-//ZZ //   LINX_(__NR_pselect6,       sys_pselect6),         //
-//ZZ 
-//ZZ //   LINX_(__NR_unshare,       sys_unshare),          // 310
-//ZZ    LINX_(__NR_set_robust_list,    sys_set_robust_list),  // 311
-//ZZ    LINXY(__NR_get_robust_list,    sys_get_robust_list),  // 312
-//ZZ //   LINX_(__NR_sync_file_range,   sys_sync_file_range),  // 314
-//ZZ 
-//ZZ    LINXY(__NR_move_pages,        sys_move_pages),       // 317
-//ZZ //   LINX_(__NR_getcpu,            sys_ni_syscall),       // 318
-//ZZ 
-//ZZ    LINXY(__NR_signalfd,          sys_signalfd),         // 321
-//ZZ    LINXY(__NR_eventfd,           sys_eventfd),          // 323
-//ZZ 
-//ZZ 
-//ZZ    ///////////////
-//ZZ 
-//ZZ    // JRS 2010-Jan-03: I believe that all the numbers listed 
-//ZZ    // in comments in the table prior to this point (eg "// 326",
-//ZZ    // etc) are bogus since it looks to me like they are copied
-//ZZ    // verbatim from syswrap-x86-linux.c and they certainly do not
-//ZZ    // correspond to what's in include/vki/vki-scnums-arm-linux.h.
-//ZZ    // From here onwards, please ensure the numbers are correct.
-//ZZ 
-//ZZ 
-//ZZ    LINXY(__NR_epoll_pwait,       sys_epoll_pwait),      // 346
-//ZZ 
-//ZZ 
-//ZZ    LINXY(__NR_eventfd2,          sys_eventfd2),         // 356
-//ZZ    LINXY(__NR_epoll_create1,     sys_epoll_create1),    // 357
-//ZZ    LINXY(__NR_rt_tgsigqueueinfo, sys_rt_tgsigqueueinfo),// 363
-//ZZ 
-//ZZ    LINXY(__NR_clock_adjtime,     sys_clock_adjtime)     // 372
+   //   (__NR_bpf,               sys_ni_syscall)         // 280
+   //   (__NR_execveat,          sys_ni_syscall),        // 281
+   //   (__NR_userfaultfd,       sys_ni_syscall),        // 282
+   //   (__NR_membarrier,        sys_ni_syscall),        // 283
+   //   (__NR_mlock2,            sys_ni_syscall),        // 284
+   //   (__NR_copy_file_range,   sys_ni_syscall),        // 285
+   //   (__NR_preadv2,           sys_ni_syscall),        // 286
+   //   (__NR_pwritev2,          sys_ni_syscall),        // 287
+   //   (__NR_pkey_mprotect,     sys_ni_syscall),        // 288
+   //   (__NR_pkey_alloc,        sys_ni_syscall),        // 289
+   //   (__NR_pkey_free,         sys_ni_syscall),        // 290
 };
 
 
