diff -Naurp a/none/tests/arm/Makefile.am b/none/tests/arm/Makefile.am
--- a/none/tests/arm/Makefile.am	2016-10-21 10:37:42.000000000 +0000
+++ b/none/tests/arm/Makefile.am	2018-03-28 14:04:54.789276446 +0000
@@ -58,7 +58,9 @@ v6intThumb_CFLAGS = $(AM_CFLAGS) -g -O0
 v6media_CFLAGS    = $(AM_CFLAGS) -g -O0 -mcpu=cortex-a8 -mthumb
 
 v8crypto_a_CFLAGS = $(AM_CFLAGS) -g -O0 -mfpu=crypto-neon-fp-armv8 -marm
+v8crypto_a_LDADD  = $(LDADD) -lm
 v8crypto_t_CFLAGS = $(AM_CFLAGS) -g -O0 -mfpu=crypto-neon-fp-armv8 -mthumb
+v8crypto_t_LDADD  = $(LDADD) -lm
 
 v8memory_a_CFLAGS = $(AM_CFLAGS) -g -O0 \
 			-march=armv8-a -mfpu=crypto-neon-fp-armv8 -marm
diff -Naurp a/none/tests/arm/Makefile.in b/none/tests/arm/Makefile.in
--- a/none/tests/arm/Makefile.in	2016-10-21 10:42:38.000000000 +0000
+++ b/none/tests/arm/Makefile.in	2018-03-28 14:05:11.089615984 +0000
@@ -184,12 +184,12 @@ v6media_LINK = $(CCLD) $(v6media_CFLAGS)
 	$(LDFLAGS) -o $@
 v8crypto_a_SOURCES = v8crypto_a.c
 v8crypto_a_OBJECTS = v8crypto_a-v8crypto_a.$(OBJEXT)
-v8crypto_a_LDADD = $(LDADD)
+v8crypto_a_LDADD = $(LDADD) -lm
 v8crypto_a_LINK = $(CCLD) $(v8crypto_a_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) \
 	$(LDFLAGS) -o $@
 v8crypto_t_SOURCES = v8crypto_t.c
 v8crypto_t_OBJECTS = v8crypto_t-v8crypto_t.$(OBJEXT)
-v8crypto_t_LDADD = $(LDADD)
+v8crypto_t_LDADD = $(LDADD) -lm
 v8crypto_t_LINK = $(CCLD) $(v8crypto_t_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) \
 	$(LDFLAGS) -o $@
 v8memory_a_SOURCES = v8memory_a.c
