diff -Naurp a/none/tests/arm/Makefile.am b/none/tests/arm/Makefile.am
--- a/none/tests/arm/Makefile.am	2017-06-15 13:41:33.000000000 +0000
+++ b/none/tests/arm/Makefile.am	2018-04-14 10:51:37.821525500 +0000
@@ -62,10 +62,14 @@ v6intThumb_CFLAGS = $(AM_CFLAGS) -g -O0
 v6media_CFLAGS    = $(AM_CFLAGS) -g -O0 -mcpu=cortex-a8 -mthumb
 
 v8crypto_a_CFLAGS = $(AM_CFLAGS) -g -O0 -mfpu=crypto-neon-fp-armv8 -marm
+v8crypto_a_LDADD  = $(LDADD) -lm
 v8crypto_t_CFLAGS = $(AM_CFLAGS) -g -O0 -mfpu=crypto-neon-fp-armv8 -mthumb
+v8crypto_t_LDADD  = $(LDADD) -lm
 
 v8fpsimd_a_CFLAGS = $(AM_CFLAGS) -g -O0 -mfpu=crypto-neon-fp-armv8 -marm
+v8fpsimd_a_LDADD  = $(LDADD) -lm
 v8fpsimd_t_CFLAGS = $(AM_CFLAGS) -g -O0 -mfpu=crypto-neon-fp-armv8 -mthumb
+v8fpsimd_t_LDADD  = $(LDADD) -lm
 
 v8memory_a_CFLAGS = $(AM_CFLAGS) -g -O0 \
 			-march=armv8-a -mfpu=crypto-neon-fp-armv8 -marm
diff -Naurp a/none/tests/arm/Makefile.in b/none/tests/arm/Makefile.in
--- a/none/tests/arm/Makefile.in	2017-06-15 13:41:33.000000000 +0000
+++ b/none/tests/arm/Makefile.in	2018-04-14 10:51:37.821525500 +0000
@@ -184,22 +184,22 @@ v6media_LINK = $(CCLD) $(v6media_CFLAGS)
 	$(LDFLAGS) -o $@
 v8crypto_a_SOURCES = v8crypto_a.c
 v8crypto_a_OBJECTS = v8crypto_a-v8crypto_a.$(OBJEXT)
-v8crypto_a_LDADD = $(LDADD)
+v8crypto_a_LDADD = $(LDADD) -lm
 v8crypto_a_LINK = $(CCLD) $(v8crypto_a_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) \
 	$(LDFLAGS) -o $@
 v8crypto_t_SOURCES = v8crypto_t.c
 v8crypto_t_OBJECTS = v8crypto_t-v8crypto_t.$(OBJEXT)
-v8crypto_t_LDADD = $(LDADD)
+v8crypto_t_LDADD = $(LDADD) -lm
 v8crypto_t_LINK = $(CCLD) $(v8crypto_t_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) \
 	$(LDFLAGS) -o $@
 v8fpsimd_a_SOURCES = v8fpsimd_a.c
 v8fpsimd_a_OBJECTS = v8fpsimd_a-v8fpsimd_a.$(OBJEXT)
-v8fpsimd_a_LDADD = $(LDADD)
+v8fpsimd_a_LDADD = $(LDADD) -lm
 v8fpsimd_a_LINK = $(CCLD) $(v8fpsimd_a_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) \
 	$(LDFLAGS) -o $@
 v8fpsimd_t_SOURCES = v8fpsimd_t.c
 v8fpsimd_t_OBJECTS = v8fpsimd_t-v8fpsimd_t.$(OBJEXT)
-v8fpsimd_t_LDADD = $(LDADD)
+v8fpsimd_t_LDADD = $(LDADD) -lm
 v8fpsimd_t_LINK = $(CCLD) $(v8fpsimd_t_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) \
 	$(LDFLAGS) -o $@
 v8memory_a_SOURCES = v8memory_a.c
