diff -Naurp a/src/registry.cxx b/src/registry.cxx
--- a/src/registry.cxx	2013-03-24 10:35:06.000000000 +0000
+++ b/src/registry.cxx	2018-04-20 12:42:53.278444405 +0000
@@ -40,8 +40,7 @@ extern "C" SAL_DLLPUBLIC_EXPORT void SAL
 	*ppEnvTypeName = CPPU_CURRENT_LANGUAGE_BINDING_NAME ":unsafe";
 } 
 
-Reference<XInterface> SAL_CALL regSettingsCreate(const Reference<XComponentContext> & context)
-	SAL_THROW((Exception)) {
+Reference<XInterface> SAL_CALL regSettingsCreate(const Reference<XComponentContext> & context) {
 	return static_cast< ::cppu::OWeakObject * >(new SettingsEventHandler(context));
 }
 
