--- doxygen-1.8.0/src/configgen.py	2012-01-29 10:12:54.000000000 -0500
+++ doxygen-1.8.0/src/configgen.py	2012-08-07 17:38:54.751696536 -0400
@@ -34,8 +34,10 @@
         if type=='bool':
         	if len(adefval)>0:
 			enabled = adefval
-		else:
-			enabled = "TRUE" if defval=='1' else "FALSE"
+		elif defval=='1':
+			enabled = "TRUE"
+		else:
+			enabled = "FALSE"
 		print "  cb = cfg->addBool("
 		print "                 \"%s\"," % (name)
 		print "                 \"%s\"," % (docC)
