diff -Naur bacula-5.2.2.old/src/dird/bacula-dir.conf.in bacula-5.2.2/src/dird/bacula-dir.conf.in
--- bacula-5.2.2.old/src/dird/bacula-dir.conf.in	2011-12-05 13:51:20.375282302 +0100
+++ bacula-5.2.2/src/dird/bacula-dir.conf.in	2011-12-05 13:54:12.809538069 +0100
@@ -15,7 +15,7 @@
 Director {                            # define myself
   Name = @basename@-dir
   DIRport = @dir_port@                # where we listen for UA connections
-  QueryFile = "@scriptdir@/query.sql"
+  QueryFile = "@sysconfdir@/query.sql"
   WorkingDirectory = "@working_dir@"
   PidDirectory = "@piddir@"
   Maximum Concurrent Jobs = 1
