diff --git a/src/lib/message.c b/src/lib/message.c
index c4b1b90..0851da8 100644
--- a/src/lib/message.c
+++ b/src/lib/message.c
@@ -60,7 +60,7 @@ utime_t daemon_start_time = 0;        /* Daemon start time */
 const char *version = VERSION " (" BDATE ")";
 const char *dist_name = DISTNAME " " DISTVER;
 int beef = BEEF;
-char my_name[30] = {0};               /* daemon name is stored here */
+char my_name[64] = {0};               /* daemon name is stored here */
 char host_name[50] = {0};             /* host machine name */
 char *exepath = (char *)NULL;
 char *exename = (char *)NULL;
