diff -Naur bacula-5.2.13.old/src/qt-console/bat.pro.in bacula-5.2.13/src/qt-console/bat.pro.in
--- bacula-5.2.13.old/src/qt-console/bat.pro.in	2013-02-19 20:21:35.000000000 +0100
+++ bacula-5.2.13/src/qt-console/bat.pro.in	2013-02-20 09:45:44.110033340 +0100
@@ -6,13 +6,13 @@
 #
 #   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
 #    
-CONFIG += qt debug @QWT@
+CONFIG += qt release @QWT@
 
 bins.path = /$(DESTDIR)@sbindir@
 bins.files = bat
 confs.path = /$(DESTDIR)@sysconfdir@
 confs.commands = ./install_conf_file
-help.path = /$(DESTDIR)@docdir@
+help.path = /$(DESTDIR)@htmldir@
 help.files = help/*.html images/status.png images/mail-message-new.png
 
 TEMPLATE     = app
diff -Naur bacula-5.2.13.old/src/qt-console/main.cpp bacula-5.2.13/src/qt-console/main.cpp
--- bacula-5.2.13.old/src/qt-console/main.cpp	2013-02-19 20:21:35.000000000 +0100
+++ bacula-5.2.13/src/qt-console/main.cpp	2013-02-20 09:46:19.556679571 +0100
@@ -40,9 +40,9 @@
 /*
  * We need Qt version 4.8.4 or later to be able to comple correctly
  */
-#if QT_VERSION < 0x040804
+#if QT_VERSION < 0x040602
 #error "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!"
-#error "You need Qt version 4.8.4 or later to build Bat"
+#error "You need Qt version 4.6.2 or later to build Bat"
 #error "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!"
 #endif
 
