diff -Naurp a/src/host.h.in b/src/host.h.in
--- a/src/host.h.in	2018-03-28 18:57:16.967091436 +0000
+++ b/src/host.h.in	2018-03-28 19:13:09.287075699 +0000
@@ -48,6 +48,8 @@
 #define HOST_OS  "i386-redhat-linux-gnu"
 #elif defined(__aarch64__)
 #define HOST_OS  "aarch64-redhat-linux-gnu"
+#else
+#define HOST_OS  "@host@@post_host@"
 #endif
 #define BACULA   "@BACULA@"
 #define DISTNAME "@DISTNAME@"
