diff --git a/src/host.h.in b/src/host.h.in
index 3fdbb05..1b3e684 100644
--- a/src/host.h.in
+++ b/src/host.h.in
@@ -30,7 +30,25 @@
  *
  */
 
-#define HOST_OS  "@host@@post_host@"
+#if defined(__x86_64__)
+#define HOST_OS  "x86_64-redhat-linux-gnu"
+#elif defined(__i386__)
+#define HOST_OS  "i386-redhat-linux-gnu"
+#elif defined(__ppc64__) || defined(__powerpc64__)
+#define HOST_OS  "powerpc64-redhat-linux-gnu"
+#elif defined(__ppc__) || defined(__powerpc__)
+#define HOST_OS  "powerpc-redhat-linux-gnu"
+#elif defined(__s390x__)
+#define HOST_OS  "s390x-ibm-linux-gnu"
+#elif defined(__s390__)
+#define HOST_OS  "s390-ibm-linux-gnu"
+#elif defined(__sparc__) && defined(__arch64__)
+#define HOST_OS  "x86_64-redhat-linux-gnu"
+#elif defined(__sparc__)
+#define HOST_OS  "i386-redhat-linux-gnu"
+#elif defined(__aarch64__)
+#define HOST_OS  "aarch64-redhat-linux-gnu"
+#endif
 #define BACULA   "@BACULA@"
 #define DISTNAME "@DISTNAME@"
 #define DISTVER  "@DISTVER@"
