From fe2609302233148b8c670417748c0f036092c3d7 Mon Sep 17 00:00:00 2001
From: Neal Gompa <ngompa13@gmail.com>
Date: Sun, 12 Mar 2017 18:20:49 -0400
Subject: [PATCH] Use --block-size with xz to make seekable xz-compressed
 images (rhbz#984704)

---
 appcreate/appliance.py | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/appcreate/appliance.py b/appcreate/appliance.py
index e489494..61f1c46 100644
--- a/appcreate/appliance.py
+++ b/appcreate/appliance.py
@@ -643,7 +643,8 @@ class ApplianceImageCreator(ImageCreator):
                 dst = "%s/%s-%s.%s" % (self._outdir, self.name, name, self.__disk_format)
 
                 if self.__compress:
-                    rc = subprocess.call(["xz", "-z", src])
+                    # Compress with xz using 16 MiB block size for seekability
+                    rc = subprocess.call(["xz", "-z", "--block-size=16777216", src])
                     if rc == 0:
                         logging.debug("compression successful")
                     if rc != 0:
-- 
2.9.3

