From bca4b0df03b2831c31ace011fdcf3c6e73c9df66 Mon Sep 17 00:00:00 2001
From: Neal Gompa <ngompa13@gmail.com>
Date: Feb 04 2017 23:39:27 +0000
Subject: Revert Fix import for compatibility with livecd-tools v24


---

diff --git a/appcreate/__init__.py b/appcreate/__init__.py
index 4bed653..0c4cebf 100644
--- a/appcreate/__init__.py
+++ b/appcreate/__init__.py
@@ -19,7 +19,7 @@
 
 from imgcreate.live import *
 from imgcreate.creator import *
-from imgcreate.dnfinst import *
+from imgcreate.yuminst import *
 from imgcreate.kickstart import *
 from imgcreate.fs import *
 from imgcreate.debug import *

