diff -up firefox-60.1.0/toolkit/components/remote/moz.build.disable-dbus-remote firefox-60.1.0/toolkit/components/remote/moz.build
--- firefox-60.1.0/toolkit/components/remote/moz.build.disable-dbus-remote	2018-06-21 09:29:35.975729500 +0200
+++ firefox-60.1.0/toolkit/components/remote/moz.build	2018-06-21 09:29:53.863631963 +0200
@@ -22,11 +22,6 @@ if 'gtk' in CONFIG['MOZ_WIDGET_TOOLKIT']
         'nsGTKRemoteService.cpp',
         'nsRemoteService.cpp',
     ]
-    if CONFIG['MOZ_ENABLE_DBUS']:
-        SOURCES += [
-            'nsDBusRemoteService.cpp',
-        ]
-        CXXFLAGS += CONFIG['MOZ_DBUS_GLIB_CFLAGS']
 
 FINAL_LIBRARY = 'xul'
 
diff -up firefox-60.1.0/toolkit/components/remote/nsRemoteService.cpp.disable-dbus-remote firefox-60.1.0/toolkit/components/remote/nsRemoteService.cpp
--- firefox-60.1.0/toolkit/components/remote/nsRemoteService.cpp.disable-dbus-remote	2018-06-19 22:35:27.000000000 +0200
+++ firefox-60.1.0/toolkit/components/remote/nsRemoteService.cpp	2018-06-21 09:30:24.949462465 +0200
@@ -6,7 +6,7 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 #include "nsGTKRemoteService.h"
-#ifdef MOZ_ENABLE_DBUS
+#if 0 // Disable DBus remote
 #include "nsDBusRemoteService.h"
 #endif
 #include "nsRemoteService.h"
@@ -34,7 +34,7 @@ NS_IMPL_ISUPPORTS(nsRemoteService,
 NS_IMETHODIMP
 nsRemoteService::Startup(const char* aAppName, const char* aProfileName)
 {
-#if defined(MOZ_ENABLE_DBUS)
+#if 0 // Disable DBus remote
     nsresult rv;
     mDBusRemoteService = new nsDBusRemoteService();
     rv = mDBusRemoteService->Startup(aAppName, aProfileName);
