diff -up firefox-60.0/memory/build/mozjemalloc.cpp.ppc-jemalloc firefox-60.0/memory/build/mozjemalloc.cpp
--- firefox-60.0/memory/build/mozjemalloc.cpp.ppc-jemalloc	2018-06-06 14:39:28.454371053 +0200
+++ firefox-60.0/memory/build/mozjemalloc.cpp	2018-06-06 14:39:40.242397928 +0200
@@ -180,7 +180,7 @@ using namespace mozilla;
 // Debug builds are opted out too, for test coverage.
 #ifndef MOZ_DEBUG
 #if !defined(__ia64__) && !defined(__sparc__) && !defined(__mips__) &&         \
-  !defined(__aarch64__)
+  !defined(__aarch64__) && !defined(__powerpc__)
 #define MALLOC_STATIC_PAGESIZE 1
 #endif
 #endif
